{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "wicked-pixel",
   "metadata": {},
   "source": [
    "# Mutation Variation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cooperative-composer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "capital-civilization",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "# # R script for looping throught the htseq outputs and merge them into one matrix\n",
    "# # Written by Ahmed Alhendi\n",
    "# Usage example:  Rscript htseq-combine_all.R {workdir} {Reacount_matrix_output_name}\n",
    "\n",
    "# args <- commandArgs(TRUE)\n",
    "# myoutname <- as.character(args[2])\n",
    "\n",
    "# ##Read files names\n",
    "# directory <- \"/research/projects/chlamydomonas/MAexpression/data/gene_count\"\n",
    "\n",
    "# sampleFiles <- grep(\"_genes_count\", list.files(directory), value=TRUE)\n",
    "# fileNames <- sub(\"_genes_count\", \"\", sampleFiles)\n",
    "# list_files <- list.files(path = directory, pattern = \"_genes_counts\")\n",
    "\n",
    "# cov <- list()\n",
    "# for (i in sampleFiles) {\n",
    "# filepath <- file.path(directory, paste(i))\n",
    "# cov[[i]] <- read.table(filepath, sep = \"\\t\", header=F, stringsAsFactors=FALSE)\n",
    "# colnames(cov[[i]]) <- c(\"ENSEMBL_GeneID\", i)\n",
    "# }\n",
    "\n",
    "# ##DataFrame\n",
    "# df <- Reduce(function(x,y) merge(x = x, y = y, by =\"ENSEMBL_GeneID\"), cov)\n",
    "# #print(sprintf(\"Exported merged table within work directory in txt and Rdata format with file name merged_%s_%s\", make.names(format(Sys.time(),\"%b_%d_%Y_%H_%M_%S_%Z\")), myoutname))\n",
    "# write.table(df,paste(directory, myoutname,\".txt\",sep=\"\"), sep=\"\\t\", quote= F, row.names = F)\n",
    "             \n",
    "# head(df)\n",
    "\n",
    "# # #LIBRARY SIZE NORMALIZATION\n",
    "# # library(\"scone\")\n",
    "# # df <- read.table('/research/projects/chlamydomonas/MAexpression/data/gene_countNA.txt', header=TRUE, row.names=\"ENSEMBL_GeneID\")\n",
    "# # new_df <- df[-c(1, 2, 3, 4, 5), ]\n",
    "# # df_matrix = data.matrix(new_df)\n",
    "# # df_UQ <- UQ_FN(df_matrix)\n",
    "# # write.table(df_UQ, paste(directory, \"_UQ_norm\", \".txt\", sep=\"\"), sep=\"\\t\", quote = F)\n",
    "# # # write.table(df_UQ, paste(directory, myoutname, \".txt\", sep=\"\"), sep=\"\\t\", quote = F, row.names=\"ENSEMBL_GeneID\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "purple-rehabilitation",
   "metadata": {},
   "source": [
    "### Keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "overhead-generation",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.figure import Figure\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "from scipy import stats\n",
    "stats.junk = lambda chisq, df: stats.chi2.sf(chisq, df)\n",
    "\n",
    "CC2344 = ['CC2344-ANC', \"CC2344-L1\", \"CC2344-L2\", \"CC2344-L3\", \"CC2344-L4\", \"CC2344-L5\", \"CC2344-L6\", \"CC2344-L7\", \"CC2344-L8\", \"CC2344-L9\", \"CC2344-L10\", \"CC2344-L11\", \"CC2344-L12\", \"CC2344-L13\", \"CC2344-L14\", \"CC2344-L15\"]\n",
    "CC2931 = [\"CC2931-ANC\", \"CC2931-L1\", \"CC2931-L2\", \"CC2931-L3\", \"CC2931-L4\", \"CC2931-L5\", \"CC2931-L6\", \"CC2931-L7\", \"CC2931-L9\", \"CC2931-L10\", \"CC2931-L11\", \"CC2931-L13\", \"CC2931-L14\", \"CC2931-L15\"]\n",
    "\n",
    "CC2344_list = [ i + '-rep1' for i in CC2344] + [ i + '-rep2' for i in CC2344] + [ i + '-rep3' for i in CC2344]\n",
    "CC2931_list = [ i + '-rep1' for i in CC2931] + [ i + '-rep2' for i in CC2931] + [ i + '-rep3' for i in CC2931]\n",
    "    \n",
    "CC2344_rep = [ i + '-rep1' for i in CC2344[1:]] + [ i + '-rep2' for i in CC2344[1:]] + [ i + '-rep3' for i in CC2344[1:]]\n",
    "CC2931_rep = [ i + '-rep1' for i in CC2931[1:]] + [ i + '-rep2' for i in CC2931[1:]] + [ i + '-rep3' for i in CC2931[1:]]\n",
    "\n",
    "dic_gen = {'CC2344-L1': 912.356113, 'CC2344-L10': 917.129696, 'CC2344-L11': 889.5859554, 'CC2344-L12': 950.0552184, 'CC2344-L13': 961.4186064,\n",
    "           'CC2344-L14': 931.447801, 'CC2344-L15': 946.6643063, 'CC2344-L2': 923.1078072, 'CC2344-L3': 1000.469526, 'CC2344-L4': 808.9505794,\n",
    "           'CC2344-L5': 957.6380465, 'CC2344-L6': 970.6307256, 'CC2344-L7': 990.9451516, 'CC2344-L8': 1009.966123, 'CC2344-L9': 901.0619061, \n",
    "           'CC2931-L1': 1050.109001, 'CC2931-L10': 1097.978141, 'CC2931-L11': 1021.13559, 'CC2931-L13': 1041.362593, \n",
    "           'CC2931-L14': 1016.111493, 'CC2931-L15': 1052.540951, 'CC2931-L2': 1056.765369, 'CC2931-L3': 1000.399127, 'CC2931-L4': 1011.411706,\n",
    "           'CC2931-L5': 993.8603657, 'CC2931-L6': 1083.095655, 'CC2931-L7': 1067.34507, 'CC2931-L9': 1079.236285}\n",
    "\n",
    "dic_mut = {'CC2344-L1': 396, 'CC2344-L10': 59, 'CC2344-L11': 46, 'CC2344-L12': 74, 'CC2344-L13': 49, 'CC2344-L14': 46, 'CC2344-L15': 53, \n",
    "           'CC2344-L2': 80, 'CC2344-L3': 63, 'CC2344-L4': 24, 'CC2344-L5': 68, 'CC2344-L6': 38, 'CC2344-L7': 45, 'CC2344-L8': 75, 'CC2344-L9': 27, \n",
    "           'CC2931-L1': 89, 'CC2931-L10': 87, 'CC2931-L11': 85, 'CC2931-L13': 97, 'CC2931-L14': 79, 'CC2931-L15': 141, 'CC2931-L2': 123,\n",
    "           'CC2931-L3': 52, 'CC2931-L4': 100, 'CC2931-L5': 335, 'CC2931-L6': 84, 'CC2931-L7': 72, 'CC2931-L9': 113}\n",
    "\n",
    "generations = pd.Series(dic_gen)\n",
    "mutations = pd.Series(dic_mut)\n",
    "\n",
    "CC2344_gen_mean = sum([dic_gen[i] for i in CC2344[1:]])/len([dic_gen[i] for i in CC2344[1:]])\n",
    "CC2931_gen_mean = sum([dic_gen[i] for i in CC2931[1:]])/len([dic_gen[i] for i in CC2931[1:]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sonic-killer",
   "metadata": {},
   "source": [
    "### Observed vs Expected Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "developed-personal",
   "metadata": {},
   "outputs": [],
   "source": [
    "CC2931_mut_chi = pd.DataFrame()\n",
    "CC2344_mut_chi = pd.DataFrame()\n",
    "\n",
    "## Based on the generation time and the known mutation rate, how many mutations do we expect to see\n",
    "# CHI-SQUARE ANALYSIS (actual vs expected mutations)\n",
    "\n",
    "for i in CC2931[1:]:\n",
    "    CC2931_mut_chi.at[i, 'exp_mut'] = generations[i]*(11.5*10**(-10))*(111*10**(6))\n",
    "    CC2931_mut_chi.at[i, 'mut'] = dic_mut[i]\n",
    "    CC2931_mut_chi.at[i, 'stat'] = (CC2931_mut_chi.at[i, 'mut'] - CC2931_mut_chi.at[i, 'exp_mut'])**2/(CC2931_mut_chi.at[i, 'exp_mut'])\n",
    "CC2931_mut_chi['chi_square'] = CC2931_mut_chi['stat'].sum()\n",
    "CC2931_mut_chi['chi_sq_min_outliers'] = CC2931_mut_chi['stat'].sum() - CC2931_mut_chi.at['CC2931-L5','exp_mut']\n",
    "CC2931_mut_chi['p-val_1'] = stats.junk(CC2931_mut_chi['chi_square'], len(CC2931_mut_chi.index.values))\n",
    "CC2931_mut_chi['p-val_2'] = stats.junk(CC2931_mut_chi['chi_sq_min_outliers'], len(CC2931_mut_chi.index.values) - 1)\n",
    "\n",
    "for i in CC2344[1:]:\n",
    "    CC2344_mut_chi.at[i, 'exp_mut'] = generations[i]*(11.5*10**(-10))*(111*10**(6))\n",
    "    CC2344_mut_chi.at[i, 'mut'] = dic_mut[i]\n",
    "    CC2344_mut_chi.at[i, 'stat'] = (CC2344_mut_chi.at[i, 'mut'] - CC2344_mut_chi.at[i, 'exp_mut'])**2/(CC2344_mut_chi.at[i, 'exp_mut'])\n",
    "CC2344_mut_chi['chi_square'] = CC2344_mut_chi['stat'].sum()\n",
    "CC2344_mut_chi['chi_sq_min_outliers'] = CC2344_mut_chi['stat'].sum() - CC2344_mut_chi.at['CC2344-L1','exp_mut']\n",
    "CC2344_mut_chi['p-val_1'] = stats.junk(CC2344_mut_chi['chi_square'], len(CC2344_mut_chi.index.values))\n",
    "CC2344_mut_chi['p-val_2'] = stats.junk(CC2344_mut_chi['chi_sq_min_outliers'], len(CC2344_mut_chi.index.values) - 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "elegant-hunger",
   "metadata": {},
   "source": [
    "### Cleaning Up The Data Provided From DESeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "looking-shell",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## CLEANING UP THE DATA PROVIDED FROM DESEQ2\n",
    "# CC2931_new_columns = []\n",
    "# CC2344_new_columns = []\n",
    "\n",
    "# CC2931_rel = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/normalized_raw_counts/gene_count_deseq_CC2931.txt', delimiter = \"\\t\")\n",
    "# CC2344_rel = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/normalized_raw_counts/gene_count_deseq_CC2344.txt', delimiter = \"\\t\")\n",
    "\n",
    "# for i in CC2931_rel.columns:\n",
    "#     i = i.replace('_', '-')\n",
    "#     for j in CC2931_list:\n",
    "#         if j in i:\n",
    "#             CC2931_new_columns.append(j)\n",
    "            \n",
    "# for i in CC2344_rel.columns:\n",
    "#     i = i.replace('_', '-')\n",
    "#     for j in CC2344_list:\n",
    "#         if j in i:\n",
    "#             CC2344_new_columns.append(j)\n",
    "\n",
    "# CC2931_rel.columns = CC2931_new_columns\n",
    "# CC2344_rel.columns = CC2344_new_columns\n",
    "# CC2931_rel.index.name = 'genes'\n",
    "# CC2344_rel.index.name = 'genes'\n",
    "# genes = list(CC2931_rel.index.values)\n",
    "\n",
    "# CC2931_rel['CC2931-ANC'] = CC2931_rel[['CC2931-ANC-rep1', 'CC2931-ANC-rep2', 'CC2931-ANC-rep3']].mean(axis = 1)\n",
    "# CC2344_rel['CC2344-ANC'] = CC2344_rel[['CC2344-ANC-rep1', 'CC2344-ANC-rep2', 'CC2344-ANC-rep3']].mean(axis = 1)\n",
    "\n",
    "# CC2931_rel.drop(['CC2931-ANC-rep1', 'CC2931-ANC-rep2', 'CC2931-ANC-rep3'], inplace=True, axis=1)\n",
    "# CC2344_rel.drop(['CC2344-ANC-rep1', 'CC2344-ANC-rep2', 'CC2344-ANC-rep3'], inplace=True, axis=1)\n",
    "\n",
    "# for i in genes:\n",
    "#     if CC2344_rel.at[i, 'CC2344-ANC'] == 0:\n",
    "#         CC2344_rel.drop(i, inplace = True)\n",
    "#     if CC2931_rel.at[i, 'CC2931-ANC'] == 0:\n",
    "#         CC2931_rel.drop(i, inplace = True)\n",
    "        \n",
    "# ## EXPORTING FILTERED RAW COUNTS\n",
    "# CC2931_rel.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/CC2931_raw', sep = \"\\t\", index = True, header = True)\n",
    "# CC2344_rel.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/CC2344_raw', sep = \"\\t\", index = True, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "legal-sphere",
   "metadata": {},
   "source": [
    "### Identifying Low and High Expression Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "warming-negotiation",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# IDENTIFYING HIGH AND LOW EXPRESSION GENES IN ANCESTORS AND MA\n",
    "CC2931_raw = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/CC2931_raw', delimiter = '\\t')\n",
    "CC2344_raw = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/CC2344_raw', delimiter = '\\t')\n",
    "CC2931_raw.set_index('genes', inplace = True)\n",
    "CC2344_raw.set_index('genes', inplace = True)\n",
    "\n",
    "for i in CC2344[1:] + CC2931[1:]:\n",
    "    if 'CC2931' in i:\n",
    "        CC2931_raw[i] = CC2931_raw[[i + '-rep1', i + '-rep2', i + '-rep3']].mean(axis = 1)\n",
    "    if 'CC2344' in i:\n",
    "        CC2344_raw[i] = CC2344_raw[[i + '-rep1', i + '-rep2', i + '-rep3']].mean(axis = 1)\n",
    "\n",
    "CC2344_10per = CC2344_raw['CC2344-ANC'].quantile(0.1)\n",
    "CC2931_10per = CC2931_raw['CC2931-ANC'].quantile(0.1)\n",
    "CC2344_90per = CC2344_raw['CC2344-ANC'].quantile(0.9)\n",
    "CC2931_90per = CC2931_raw['CC2931-ANC'].quantile(0.9)\n",
    "\n",
    "CC2344_low_exp = list(CC2344_raw.loc[CC2344_raw['CC2344-ANC'] <= CC2344_10per].index.values)\n",
    "CC2931_low_exp = list(CC2931_raw.loc[CC2931_raw['CC2931-ANC'] <= CC2931_10per].index.values)\n",
    "\n",
    "CC2344_hi_exp = list(CC2344_raw.loc[CC2344_raw['CC2344-ANC'] >= CC2344_90per].index.values)\n",
    "CC2931_hi_exp = list(CC2931_raw.loc[CC2931_raw['CC2931-ANC'] >= CC2931_90per].index.values)\n",
    "\n",
    "HL_MAgenes = {}\n",
    "for i in CC2344[1:] + CC2931[1:]:\n",
    "    if 'CC2344' in i:\n",
    "        HL_MAgenes[i + \"_low\"] = list(CC2344_raw.loc[CC2344_raw[i] <= CC2344_10per].index.values)\n",
    "        HL_MAgenes[i + \"_hi\"] = list(CC2344_raw.loc[CC2344_raw[i] >= CC2344_90per].index.values)\n",
    "    if 'CC2931' in i:\n",
    "        HL_MAgenes[i + \"_low\"] = list(CC2931_raw.loc[CC2931_raw[i] <= CC2931_10per].index.values)\n",
    "        HL_MAgenes[i + \"_hi\"] = list(CC2931_raw.loc[CC2931_raw[i] >= CC2931_90per].index.values)\n",
    "\n",
    "## CHECKING INTERSECTION SETS OF HIGH AND LOW EXPRESSION FOR EACH GENOTYPE\n",
    "verify_hi_hi = list(set(CC2344_hi_exp).symmetric_difference(set(CC2931_hi_exp)))\n",
    "verify_hi_low = list(set.intersection(set(CC2344_hi_exp), set(CC2931_low_exp))) #noticed points of intersection between sites of high and low expression\n",
    "verify_low_hi = list(set.intersection(set(CC2344_low_exp), set(CC2931_hi_exp)))\n",
    "verify_low_low = list(set(CC2344_low_exp).symmetric_difference(set(CC2931_low_exp)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "emerging-assessment",
   "metadata": {},
   "source": [
    "### Relative Fold Change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ranging-diving",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.3714873813843594e-287 3.1836735487880585e-242 0.0 0.0\n"
     ]
    }
   ],
   "source": [
    "## TAKING THE RELATIVE FOLD CHANGE\n",
    "for i in list(CC2344_rel.columns) + list(CC2931_rel.columns):\n",
    "    if 'CC2931' in i:\n",
    "        CC2931_rel[i] = CC2931_rel[i]/CC2931_rel['CC2931-ANC']\n",
    "    if 'CC2344' in i:\n",
    "        CC2344_rel[i] = CC2344_rel[i]/CC2344_rel['CC2344-ANC']\n",
    "\n",
    "for i in CC2344[1:] + CC2931[1:]:\n",
    "    if 'CC2931' in i:\n",
    "        CC2931_rel[i] = CC2931_rel[[i + '-rep1', i + '-rep2', i + '-rep3']].mean(axis = 1)\n",
    "    if 'CC2344' in i:\n",
    "        CC2344_rel[i] = CC2344_rel[[i + '-rep1', i + '-rep2', i + '-rep3']].mean(axis = 1)\n",
    "\n",
    "## FINDING THE RELATIVE FOLD CHANGE OVER MA GENERATIONS (VARIANCE PER GENE)\n",
    "CC2931_rel_per_gen = pd.DataFrame()\n",
    "CC2344_rel_per_gen = pd.DataFrame()\n",
    "\n",
    "CC2344_rel_per_gen['mut_variance'] = CC2344_rel.var(axis = 1)/CC2344_gen_mean\n",
    "CC2931_rel_per_gen['mut_variance'] = CC2931_rel.var(axis = 1)/CC2931_gen_mean\n",
    "\n",
    "## EXPORTING AS CSV\n",
    "CC2931_rel_per_gen.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/relative_fold_change/CC2931_pergen', sep = \"\\t\", header = True, index = True)\n",
    "CC2344_rel_per_gen.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/relative_fold_change/CC2344_pergen', sep = \"\\t\", header = True, index = True)\n",
    "\n",
    "## FINDING THE VARIANCE OF THE RELATIVE FOLD MA LINES FROM ANCESTOR\n",
    "CC2931_rel_var = pd.DataFrame()\n",
    "CC2344_rel_var = pd.DataFrame()\n",
    "\n",
    "for i in CC2931[1:] + CC2344[1:]:\n",
    "    if 'CC2931' in i:\n",
    "        CC2931_rel_var[i + '_dif_ANC'] = (CC2931_rel[i] - CC2931_rel['CC2931-ANC'])**2\n",
    "        CC2931_rel_var['sum'] = CC2931_rel_var[CC2931_rel_var.columns].sum(axis = 1)\n",
    "        CC2931_rel_var['variance'] = CC2931_rel_var['sum']/13\n",
    "        CC2931_rel_var['mut_variance'] = CC2931_rel_var['variance']/CC2931_gen_mean\n",
    "    if 'CC2344' in i:\n",
    "        CC2344_rel_var[i + '_dif_ANC'] = (CC2344_rel[i] - CC2344_rel['CC2344-ANC'])**2\n",
    "        CC2344_rel_var['sum'] = CC2344_rel_var[CC2344_rel_var.columns].sum(axis = 1)\n",
    "        CC2344_rel_var['variance'] = CC2344_rel_var['sum']/15\n",
    "        CC2344_rel_var['mut_variance'] = CC2344_rel_var['variance']/CC2344_gen_mean\n",
    "        \n",
    "############################################################################################################################\n",
    "from scipy.stats import mannwhitneyu\n",
    "for c in ['CC2344_high', 'CC2344_low', 'CC2344_high1', 'CC2344_low1', 'CC2931_high', 'CC2931_low', 'CC2931_high1', 'CC2931_low1']:\n",
    "     exec('{} = pd.DataFrame()'.format(c))\n",
    "\n",
    "## RELATIVE FOLD VARIANCE PER GENERATION\n",
    "CC2344_high['mut_variance'] = CC2344_rel_per_gen['mut_variance'].loc[CC2344_hi_exp].values.tolist()\n",
    "CC2344_high['expression'] = 'high'\n",
    "CC2344_high['id'] = 'CC2344'\n",
    "CC2344_low['mut_variance'] = CC2344_rel_per_gen['mut_variance'].loc[CC2344_low_exp].values.tolist()\n",
    "CC2344_low['expression'] = 'low'\n",
    "CC2344_low['id'] = 'CC2344'\n",
    "\n",
    "CC2931_high['mut_variance'] = CC2931_rel_per_gen['mut_variance'].loc[CC2931_hi_exp].values.tolist()\n",
    "CC2931_high['expression'] = 'high'\n",
    "CC2931_high['id'] = 'CC2931'\n",
    "CC2931_low['mut_variance'] = CC2931_rel_per_gen['mut_variance'].loc[CC2931_low_exp].values.tolist()\n",
    "CC2931_low['expression'] = 'low'\n",
    "CC2931_low['id'] = 'CC2931'\n",
    "\n",
    "rel_per_gen = pd.concat([CC2344_high, CC2344_low, CC2931_high, CC2931_low], axis = 0)\n",
    "\n",
    "## RELATIVE FOLD VARIANCE FROM ANCESTOR\n",
    "CC2344_high1['mut_variance'] = CC2344_rel_var['mut_variance'].loc[CC2344_hi_exp].values.tolist()\n",
    "CC2344_high1['expression'] = 'high'\n",
    "CC2344_high1['id'] = 'CC2344'\n",
    "CC2344_low1['mut_variance'] = CC2344_rel_var['mut_variance'].loc[CC2344_low_exp].values.tolist()\n",
    "CC2344_low1['expression'] = 'low'\n",
    "CC2344_low1['id'] = 'CC2344'\n",
    "\n",
    "CC2931_high1['mut_variance'] = CC2931_rel_var['mut_variance'].loc[CC2931_hi_exp].values.tolist()\n",
    "CC2931_high1['expression'] = 'high'\n",
    "CC2931_high1['id'] = 'CC2931'\n",
    "CC2931_low1['mut_variance'] = CC2931_rel_var['mut_variance'].loc[CC2931_low_exp].values.tolist()\n",
    "CC2931_low1['expression'] = 'low'\n",
    "CC2931_low1['id'] = 'CC2931'\n",
    "\n",
    "rel_fold_change = pd.concat([CC2344_high1, CC2344_low1, CC2931_high1, CC2931_low1], axis = 0)\n",
    "\n",
    "## MANN WHITNEY U TEST\n",
    "U1, p1 = mannwhitneyu(CC2344_high['mut_variance'].values.tolist(), CC2344_low['mut_variance'].values.tolist())\n",
    "U2, p2 = mannwhitneyu(CC2344_high1['mut_variance'].values.tolist(), CC2344_low1['mut_variance'].values.tolist())\n",
    "U3, p3 = mannwhitneyu(CC2931_high['mut_variance'].values.tolist(), CC2931_low['mut_variance'].values.tolist())\n",
    "U4, p4 = mannwhitneyu(CC2931_high1['mut_variance'].values.tolist(), CC2931_low1['mut_variance'].values.tolist())\n",
    "print(p1, p2, p3, p4)\n",
    "\n",
    "# # SEABORN BOXPLOTS\n",
    "# sns.boxplot(data = rel_per_gen, x = 'expression', y = 'mut_variance', hue = 'id')\n",
    "# sns.boxplot(data = rel_fold_change, x = 'expression', y = 'mut_variance', hue = 'id')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "numeric-sarah",
   "metadata": {},
   "source": [
    "### Absolute Fold Change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "studied-sigma",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0 0.0 0.0 0.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/balogu15/.conda/envs/personal/lib/python3.8/site-packages/scipy/stats/stats.py:7171: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  z = (bigu - meanrank) / sd\n"
     ]
    }
   ],
   "source": [
    "CC2931_abs = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/CC2931_raw', delimiter = '\\t')\n",
    "CC2344_abs = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/CC2344_raw', delimiter = '\\t')\n",
    "CC2931_abs.set_index('genes', inplace = True)\n",
    "CC2344_abs.set_index('genes', inplace = True)\n",
    "CC2931_abs_per_gen = pd.DataFrame()\n",
    "CC2344_abs_per_gen = pd.DataFrame()\n",
    "\n",
    "for i in CC2344_rep + CC2931_rep + ['CC2931-ANC', 'CC2344-ANC']:\n",
    "    if 'CC2931' in i:\n",
    "        CC2931_abs[i] = CC2931_abs[i]/CC2931_abs['CC2931-ANC'].median()\n",
    "    if 'CC2344' in i:\n",
    "        CC2344_abs[i] = CC2344_abs[i]/CC2344_abs['CC2344-ANC'].median()\n",
    "        \n",
    "# TAKING THE MEAN OF THREE REPS\n",
    "for i in CC2344[1:] + CC2931[1:]:\n",
    "    if 'CC2931' in i:\n",
    "        CC2931_abs[i] = CC2931_abs[[i + '-rep1', i + '-rep2', i + '-rep3']].mean(axis = 1)\n",
    "    if 'CC2344' in i:\n",
    "        CC2344_abs[i] = CC2344_abs[[i + '-rep1', i + '-rep2', i + '-rep3']].mean(axis = 1)\n",
    "\n",
    "# TAKING THE VARIANCE OF THE MA LINES ACROSS EACH ROW (PER GENE)\n",
    "CC2931_abs_per_gen['mut_variance'] = CC2931_abs_per_gen.var(axis = 1)/CC2931_gen_mean\n",
    "CC2344_abs_per_gen['mut_variance'] = CC2344_abs_per_gen.var(axis = 1)/CC2344_gen_mean\n",
    "\n",
    "# EXPORTING AS CSV\n",
    "CC2931_abs_per_gen.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/absolute_fold_change/CC2931_pergen', sep = \"\\t\", header = True, index = True)\n",
    "CC2344_abs_per_gen.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/absolute_fold_change/CC2344_pergen', sep = \"\\t\", header = True, index = True)\n",
    "\n",
    "## FINDING THE VARIANCE OF THE RELATIVE FOLD MA LINES FROM ANCESTOR\n",
    "CC2931_abs_var = pd.DataFrame()\n",
    "CC2344_abs_var = pd.DataFrame()\n",
    "\n",
    "for i in CC2931[1:] + CC2344[1:]:\n",
    "    if 'CC2931' in i:\n",
    "        CC2931_abs_var[i + '_dif_ANC'] = (CC2931_abs[i] - CC2931_abs['CC2931-ANC'])**2\n",
    "        CC2931_abs_var['sum'] = CC2931_abs_var[CC2931_abs_var.columns].sum(axis = 1)\n",
    "        CC2931_abs_var['variance'] = CC2931_abs_var['sum']/13\n",
    "        CC2931_abs_var['mut_variance'] = CC2931_abs_var['variance']/CC2931_gen_mean\n",
    "    if 'CC2344' in i:\n",
    "        CC2344_abs_var[i + '_dif_ANC'] = (CC2344_abs[i] - CC2344_abs['CC2344-ANC'])**2\n",
    "        CC2344_abs_var['sum'] = CC2344_abs_var[CC2344_abs_var.columns].sum(axis = 1)\n",
    "        CC2344_abs_var['variance'] = CC2344_abs_var['sum']/15\n",
    "        CC2344_abs_var['mut_variance'] = CC2344_abs_var['variance']/CC2344_gen_mean\n",
    "############################################################################################################################\n",
    "for c in ['CC2344_high', 'CC2344_low', 'CC2344_high1', 'CC2344_low1', 'CC2931_high', 'CC2931_low', 'CC2931_high1', 'CC2931_low1']:\n",
    "     exec('{} = pd.DataFrame()'.format(c))\n",
    "        \n",
    "# intersection_set = list(set.intersection(set(CC2344_hi_exp), set(CC2344_abs_per_gen.index.values)))\n",
    "\n",
    "## ABSOLUTE FOLD VARIANCE PER GENERATION\n",
    "CC2344_high['mut_variance'] = CC2344_abs_per_gen['mut_variance'].loc[list(set.intersection(set(CC2344_hi_exp), set(CC2344_abs_per_gen.index.values)))].values.tolist()\n",
    "CC2344_high['expression'] = 'high'\n",
    "CC2344_high['id'] = 'CC2344'\n",
    "CC2344_low['mut_variance'] = CC2344_abs_per_gen['mut_variance'].loc[list(set.intersection(set(CC2344_low_exp), set(CC2344_abs_per_gen.index.values)))].values.tolist()\n",
    "CC2344_low['expression'] = 'low'\n",
    "CC2344_low['id'] = 'CC2344'\n",
    "\n",
    "CC2931_high['mut_variance'] = CC2931_abs_per_gen['mut_variance'].loc[list(set.intersection(set(CC2931_hi_exp), set(CC2931_abs_per_gen.index.values)))].values.tolist()\n",
    "CC2931_high['expression'] = 'high'\n",
    "CC2931_high['id'] = 'CC2931'\n",
    "CC2931_low['mut_variance'] = CC2931_abs_per_gen['mut_variance'].loc[list(set.intersection(set(CC2931_low_exp), set(CC2931_abs_per_gen.index.values)))].values.tolist()\n",
    "CC2931_low['expression'] = 'low'\n",
    "CC2931_low['id'] = 'CC2931'\n",
    "\n",
    "abs_per_gen = pd.concat([CC2344_high, CC2344_low, CC2931_high, CC2931_low], axis = 0)\n",
    "\n",
    "## ABSOLUTE FOLD VARIANCE FROM ANCESTOR\n",
    "CC2344_high1['mut_variance'] = CC2344_abs_var['mut_variance'].loc[list(set.intersection(set(CC2344_hi_exp), set(CC2344_abs_var.index.values)))].values.tolist()\n",
    "CC2344_high1['expression'] = 'high'\n",
    "CC2344_high1['id'] = 'CC2344'\n",
    "CC2344_low1['mut_variance'] = CC2344_abs_var['mut_variance'].loc[list(set.intersection(set(CC2344_low_exp), set(CC2344_abs_var.index.values)))].values.tolist()\n",
    "CC2344_low1['expression'] = 'low'\n",
    "CC2344_low1['id'] = 'CC2344'\n",
    "\n",
    "CC2931_high1['mut_variance'] = CC2931_abs_var['mut_variance'].loc[list(set.intersection(set(CC2931_hi_exp), set(CC2931_abs_var.index.values)))].values.tolist()\n",
    "CC2931_high1['expression'] = 'high'\n",
    "CC2931_high1['id'] = 'CC2931'\n",
    "CC2931_low1['mut_variance'] = CC2931_abs_var['mut_variance'].loc[list(set.intersection(set(CC2931_low_exp), set(CC2931_abs_var.index.values)))].values.tolist()\n",
    "CC2931_low1['expression'] = 'low'\n",
    "CC2931_low1['id'] = 'CC2931'\n",
    "\n",
    "abs_fold_change = pd.concat([CC2344_high, CC2344_low, CC2931_high, CC2931_low], axis = 0)\n",
    "\n",
    "## MANN WHITNEY U TEST\n",
    "U1, p1 = mannwhitneyu(CC2344_high['mut_variance'].values.tolist(), CC2344_low['mut_variance'].values.tolist())\n",
    "U2, p2 = mannwhitneyu(CC2344_high1['mut_variance'].values.tolist(), CC2344_low1['mut_variance'].values.tolist())\n",
    "U3, p3 = mannwhitneyu(CC2931_high['mut_variance'].values.tolist(), CC2931_low['mut_variance'].values.tolist())\n",
    "U4, p4 = mannwhitneyu(CC2931_high1['mut_variance'].values.tolist(), CC2931_low1['mut_variance'].values.tolist())\n",
    "print(p1, p2, p3, p4)\n",
    "\n",
    "# ## SEABORN\n",
    "# sns.boxplot(data = abs_per_gen, x = 'expression', y = 'mut_variance', hue = 'id')\n",
    "# sns.boxplot(data = abs_fold_change, x = 'expression', y = 'mut_variance', hue = 'id')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "strategic-nashville",
   "metadata": {},
   "source": [
    "### Log2Fold - Mutational Variance (Vm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "cloudy-polyester",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7.330221812082584e-175 1.3440365276803627e-235 0.0 6.8554571747739e-247\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "CC2931_log2Fold = pd.DataFrame()\n",
    "CC2344_log2Fold = pd.DataFrame()\n",
    "combined_log2Fold = pd.DataFrame()\n",
    "\n",
    "filenames = CC2344[1:] + CC2931[1:]\n",
    "\n",
    "## THE MEAN OF THE GENERATIONS ACROSS MA LINES IN EACH GENOTYPE\n",
    "combined_gen_mean = sum([dic_gen[i] for i in CC2344[1:] + CC2931[1:]])/len([dic_gen[i] for i in CC2344[1:] + CC2931[1:]])\n",
    "\n",
    "for i in CC2931[1:]:\n",
    "    CC2931_log2Fold[i] = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/genes_log2fold/' + i, delimiter = \"\\t\")['log2FoldChange']\n",
    "for i in CC2344[1:]:\n",
    "    CC2344_log2Fold[i] = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/genes_log2fold/' + i, delimiter = \"\\t\")['log2FoldChange']\n",
    "for i in CC2931[1:] + CC2344[1:]:\n",
    "    combined_log2Fold[i] = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/genes_log2fold/' + i, delimiter = '\\t')['log2FoldChange']\n",
    "\n",
    "CC2344_log2Fold['mut_variance'] = CC2344_log2Fold.var(axis = 1)/CC2344_gen_mean\n",
    "CC2931_log2Fold['mut_variance'] = CC2931_log2Fold.var(axis = 1)/CC2931_gen_mean\n",
    "combined_log2Fold['mut_variance'] = combined_log2Fold.var(axis = 1)/combined_gen_mean\n",
    "\n",
    "############################################################################################################################\n",
    "for c in ['CC2344_high', 'CC2344_low', 'CC2344_high1', 'CC2344_low1', 'CC2931_high', 'CC2931_low', 'CC2931_high1', 'CC2931_low1']:\n",
    "     exec('{} = pd.DataFrame()'.format(c))\n",
    "\n",
    "## LOG2FOLD VARIANCE PER AVERAGE GENERATION\n",
    "CC2344_high['mut_variance'] = CC2344_log2Fold['mut_variance'].loc[CC2344_hi_exp].values.tolist()\n",
    "CC2344_high['expression'] = 'high'\n",
    "CC2344_high['id'] = 'CC2344'\n",
    "CC2344_low['mut_variance'] = CC2344_log2Fold['mut_variance'].loc[CC2344_low_exp].values.tolist()\n",
    "CC2344_low['expression'] = 'low'\n",
    "CC2344_low['id'] = 'CC2344'\n",
    "\n",
    "CC2931_high['mut_variance'] = CC2931_log2Fold['mut_variance'].loc[CC2931_hi_exp].values.tolist()\n",
    "CC2931_high['expression'] = 'high'\n",
    "CC2931_high['id'] = 'CC2931'\n",
    "CC2931_low['mut_variance'] = CC2931_log2Fold['mut_variance'].loc[CC2931_low_exp].values.tolist()\n",
    "CC2931_low['expression'] = 'low'\n",
    "CC2931_low['id'] = 'CC2931'\n",
    "\n",
    "CC2344_high_median = CC2344_high['mut_variance'].median(axis = 0, skipna = True)\n",
    "CC2344_low_median = CC2344_low['mut_variance'].median(axis = 0, skipna = True)\n",
    "CC2931_high_median = CC2931_high['mut_variance'].median(axis = 0, skipna = True)\n",
    "CC2931_low_median = CC2931_low['mut_variance'].median(axis = 0, skipna = True)\n",
    "\n",
    "log2fold_change = pd.concat([CC2344_high, CC2344_low, CC2931_high, CC2931_low], axis = 0)\n",
    "\n",
    "## COMBINED LOG2FOLD VARIANCE PER AVERAGE GENERATION\n",
    "CC2344_high1['mut_variance'] = combined_log2Fold['mut_variance'].loc[CC2344_hi_exp].values.tolist()\n",
    "CC2344_high1['expression'] = 'high'\n",
    "CC2344_high1['id'] = 'CC2344'\n",
    "CC2344_low1['mut_variance'] = combined_log2Fold['mut_variance'].loc[CC2344_low_exp].values.tolist()\n",
    "CC2344_low1['expression'] = 'low'\n",
    "CC2344_low1['id'] = 'CC2344'\n",
    "\n",
    "CC2931_high1['mut_variance'] = combined_log2Fold['mut_variance'].loc[CC2931_hi_exp].values.tolist()\n",
    "CC2931_high1['expression'] = 'high'\n",
    "CC2931_high1['id'] = 'CC2931'\n",
    "CC2931_low1['mut_variance'] = combined_log2Fold['mut_variance'].loc[CC2931_low_exp].values.tolist()\n",
    "CC2931_low1['expression'] = 'low'\n",
    "CC2931_low1['id'] = 'CC2931'\n",
    "\n",
    "CC2344_high_median = CC2344_high1['mut_variance'].median(axis = 0, skipna = True)\n",
    "CC2344_low_median = CC2344_low1['mut_variance'].median(axis = 0, skipna = True)\n",
    "CC2931_high_median = CC2931_high1['mut_variance'].median(axis = 0, skipna = True)\n",
    "CC2931_low_median = CC2931_low1['mut_variance'].median(axis = 0, skipna = True)\n",
    "\n",
    "combined_log2fold_change = pd.concat([CC2344_high, CC2344_low, CC2931_high, CC2931_low], axis = 0)\n",
    "\n",
    "## MANN WHITNEY U TEST\n",
    "U1, p1 = mannwhitneyu(CC2344_high['mut_variance'].values.tolist(), CC2344_low['mut_variance'].values.tolist())\n",
    "U2, p2 = mannwhitneyu(CC2344_high1['mut_variance'].values.tolist(), CC2344_low1['mut_variance'].values.tolist())\n",
    "U3, p3 = mannwhitneyu(CC2931_high['mut_variance'].values.tolist(), CC2931_low['mut_variance'].values.tolist())\n",
    "U4, p4 = mannwhitneyu(CC2931_high1['mut_variance'].values.tolist(), CC2931_low1['mut_variance'].values.tolist())\n",
    "print(p1, p2, p3, p4)\n",
    "\n",
    "# ## SEABORN\n",
    "# sns.boxplot(data = log2fold_change, x = 'expression', y = 'mut_variance', hue = 'id')\n",
    "# sns.boxplot(data = combined_log2fold_change, x = 'expression', y = 'mut_variance')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "resident-sterling",
   "metadata": {},
   "source": [
    "### Identifying cis mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "friendly-israel",
   "metadata": {},
   "outputs": [],
   "source": [
    "## COMBINING ALL DEGs INTO THE SAME DATAFRAME\n",
    "\n",
    "# import pandas as pd\n",
    "# import gffpandas.gffpandas as gffpd\n",
    "\n",
    "# adj_pval = pd.DataFrame()\n",
    "\n",
    "# genes = ['CC2344.resSig_L1.genes', 'CC2344.resSig_L2.genes', 'CC2344.resSig_L3.genes', 'CC2344.resSig_L4.genes',\n",
    "#            'CC2344.resSig_L5.genes', 'CC2344.resSig_L6.genes', 'CC2344.resSig_L7.genes', 'CC2344.resSig_L8.genes',\n",
    "#            'CC2344.resSig_L9.genes', 'CC2344.resSig_L10.genes', 'CC2344.resSig_L11.genes', 'CC2344.resSig_L12.genes',\n",
    "#            'CC2344.resSig_L13.genes', 'CC2344.resSig_L14.genes', 'CC2344.resSig_L15.genes', 'CC2931.resSig_L1.genes',\n",
    "#            'CC2931.resSig_L2.genes', 'CC2931.resSig_L3.genes', 'CC2931.resSig_L4.genes', 'CC2931.resSig_L5.genes',\n",
    "#            'CC2931.resSig_L6.genes', 'CC2931.resSig_L7.genes', 'CC2931.resSig_L9.genes', 'CC2931.resSig_L10.genes',\n",
    "#            'CC2931.resSig_L11.genes', 'CC2931.resSig_L13.genes', 'CC2931.resSig_L14.genes', 'CC2931.resSig_L15.genes']\n",
    "\n",
    "# for i in genes:\n",
    "#     loc = '/research/projects/chlamydomonas/MAexpression/analysis/DEGs/' + i\n",
    "#     DF = pd.read_csv(loc, delimiter = \",\")\n",
    "#     DF[[i]] = DF[['Unnamed: 0']]\n",
    "#     adj_pval = pd.concat([adj_pval, DF[i]], axis = 1)\n",
    "#     adj_pval[i] = adj_pval[i].fillna('')\n",
    "\n",
    "# adj_pval.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/DEGs/total_genes.csv', header = True, index = False)\n",
    "\n",
    "#############################################################################################################################\n",
    "\n",
    "## ISOLATING THE LOCATIONS OF DEGs\n",
    "\n",
    "# annotation = gffpd.read_gff3('/research/projects/chlamydomonas/MAexpression/data/genome_info/CC4532.v1_1.genes.primaryTrs.gff3')\n",
    "# annotation_df = annotation.df\n",
    "# annotation_df['attributes'] = annotation_df['attributes'].str.split(';', expand = True)\n",
    "\n",
    "# total_genes = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/DEGs/total_genes.csv')\n",
    "\n",
    "# CC2344 = ['CC2344-ANC', \"CC2344-L1\", \"CC2344-L2\", \"CC2344-L3\", \"CC2344-L4\", \"CC2344-L5\", \"CC2344-L6\", \"CC2344-L7\", \"CC2344-L8\", \"CC2344-L9\", \"CC2344-L10\", \"CC2344-L11\", \"CC2344-L12\", \"CC2344-L13\", \"CC2344-L14\", \"CC2344-L15\"]\n",
    "# CC2931 = [\"CC2931-ANC\", \"CC2931-L1\", \"CC2931-L2\", \"CC2931-L3\", \"CC2931-L4\", \"CC2931-L5\", \"CC2931-L6\", \"CC2931-L7\", \"CC2931-L9\", \"CC2931-L10\", \"CC2931-L11\", \"CC2931-L13\", \"CC2931-L14\", \"CC2931-L15\"]\n",
    "# list_noANC = CC2344[1:] + CC2931[1:]\n",
    "# dic = {list_noANC[i]:genes[i] for i in range(len(genes))}\n",
    "\n",
    "# for i in genes:\n",
    "#     gene_list = total_genes[i].values.tolist()\n",
    "#     gene_list = [x for x in gene_list if pd.isnull(x) == False]\n",
    "#     feature_by_attribute = annotation.get_feature_by_attribute('ID', gene_list)\n",
    "#     feature_by_attribute.to_csv(i + '.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "italian-listing",
   "metadata": {},
   "outputs": [],
   "source": [
    "## OPENING FILE WITH MUTATIONS AND THEIR KNOWN LOCATIONS IN V6\n",
    "mutations = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/mutation_info/v6_coordinates.bed', delimiter = '\\t', names = ['chromosome', 'start', 'end', 'sample'])\n",
    "mutations = mutations.sort_values('sample') # There are two missing samples CC2931_8 and CC2931_12, the first one has no recorded mutations\n",
    "\n",
    "## REMOVING ALL CC2931_12 MUTATIONS FROM THE DATAFRAME\n",
    "missing_CC2931_12 = mutations.loc[mutations['sample'] == 'CC2931_12'].index.values\n",
    "mutations.drop(missing_CC2931_12, inplace = True)\n",
    "\n",
    "## CREATING A FUNCTION THAT INDENTIFIES POSSIBLE CIS MUTATIONS\n",
    "def cis_mutations(mutations, bin_width):\n",
    "    '''this function provides a dataframe containing sites that are within a certain distance from each mutation'''\n",
    "    \n",
    "    CC2344 = [\"CC2344_1\", \"CC2344_2\", \"CC2344_3\", \"CC2344_4\", \"CC2344_5\", \"CC2344_6\", \"CC2344_7\", \"CC2344_8\", \"CC2344_9\", \"CC2344_10\", \"CC2344_11\", \"CC2344_12\", \"CC2344_13\", \"CC2344_14\", \"CC2344_15\"]\n",
    "    CC2931 = [\"CC2931_1\", \"CC2931_2\", \"CC2931_3\", \"CC2931_4\", \"CC2931_5\", \"CC2931_6\", \"CC2931_7\", \"CC2931_9\", \"CC2931_10\", \"CC2931_11\", \"CC2931_13\", \"CC2931_14\", \"CC2931_15\"]\n",
    "    total = CC2344 + CC2931\n",
    "    genes = ['CC2344.resSig_L1.genes', 'CC2344.resSig_L2.genes', 'CC2344.resSig_L3.genes', 'CC2344.resSig_L4.genes',\n",
    "           'CC2344.resSig_L5.genes', 'CC2344.resSig_L6.genes', 'CC2344.resSig_L7.genes', 'CC2344.resSig_L8.genes',\n",
    "           'CC2344.resSig_L9.genes', 'CC2344.resSig_L10.genes', 'CC2344.resSig_L11.genes', 'CC2344.resSig_L12.genes',\n",
    "           'CC2344.resSig_L13.genes', 'CC2344.resSig_L14.genes', 'CC2344.resSig_L15.genes', 'CC2931.resSig_L1.genes',\n",
    "           'CC2931.resSig_L2.genes', 'CC2931.resSig_L3.genes', 'CC2931.resSig_L4.genes', 'CC2931.resSig_L5.genes',\n",
    "           'CC2931.resSig_L6.genes', 'CC2931.resSig_L7.genes', 'CC2931.resSig_L9.genes', 'CC2931.resSig_L10.genes',\n",
    "           'CC2931.resSig_L11.genes', 'CC2931.resSig_L13.genes', 'CC2931.resSig_L14.genes', 'CC2931.resSig_L15.genes']\n",
    "    dic = {total[i]:genes[i] for i in range(len(genes))}\n",
    "\n",
    "    cis_mutations = pd.DataFrame()\n",
    "    for i in list(mutations.index.values):\n",
    "        DEGs = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/DEGs/degs_with_locations/' + dic[mutations.at[i, 'sample']] + '.csv')\n",
    "        DEGs['sample'] = mutations.at[i, 'sample']\n",
    "        mut = mutations.at[i, 'start']\n",
    "        start = mut - bin_width\n",
    "        end = mut + bin_width\n",
    "        spec_chrom = DEGs.loc[DEGs['seq_id'] == mutations.at[i,'chromosome']]\n",
    "        for a in list(spec_chrom.index.values):\n",
    "            spec_start = spec_chrom.at[a, 'start']\n",
    "            spec_end = spec_chrom.at[a, 'end']\n",
    "            if start <= spec_start and end >= spec_start:\n",
    "                cis_mutations = cis_mutations.append(spec_chrom.loc[spec_chrom['start'] == spec_start])\n",
    "            else:\n",
    "                if start <= spec_end and end >= spec_end:\n",
    "                    cis_mutations = cis_mutations.append(spec_chrom.loc[spec_chrom['start'] == spec_start])\n",
    "    return cis_mutations\n",
    "\n",
    "## FREQUENCY COUNTS OF NUMBER OF DEGs WITHIN GIVEN BANDWIDTH OF MUTATION\n",
    "frequency_table = pd.DataFrame()\n",
    "possible_cis = pd.DataFrame()\n",
    "bin_width = [10, 50, 100, 150, 200, 250, 300, 350, 400]\n",
    "for i in bin_width:\n",
    "    results = cis_mutations(mutations, i)\n",
    "    results['bin_width'] = i\n",
    "    possible_cis = possible_cis.append(results)\n",
    "    frequency_table.at[i, 'cum_counts'] = len(results)\n",
    "    \n",
    "bin_width = list(frequency_table.index.values)\n",
    "length = [i for i in range(len(frequency_table.index.values))]\n",
    "for i in length[1:]:\n",
    "    frequency_table.at[bin_width[0], 'add_counts'] = frequency_table.at[bin_width[0], 'cum_counts']\n",
    "    frequency_table.at[bin_width[i], 'add_counts'] = frequency_table.at[bin_width[i], 'cum_counts'] - frequency_table.at[bin_width[i-1],'cum_counts']\n",
    "frequency_table = frequency_table.reset_index()\n",
    "\n",
    "## CLASSIFYING POTENTIAL CIS MUTATIONS WITH HIGH/LOW/NEITHER EXPRESSION\n",
    "possible_cis['attributes'] = possible_cis['attributes'].str.replace('ID=', '', regex = True)\n",
    "possible_cis['sample'] = possible_cis['sample'].str.replace('_', '-L', regex = True)\n",
    "possible_cis['expression'] = 'neither'\n",
    "possible_cis.drop_duplicates(keep='first',inplace=True)\n",
    "possible_cis = possible_cis.reset_index()\n",
    "for i in list(possible_cis.index.values):\n",
    "    if possible_cis.at[i, 'attributes'] in HL_MAgenes[possible_cis.at[i, 'sample'] + \"_low\"]:\n",
    "        possible_cis.at[i, 'expression'] = 'low'\n",
    "    else:\n",
    "        if possible_cis.at[i, 'attributes'] in HL_MAgenes[possible_cis.at[i, 'sample'] + \"_hi\"]:\n",
    "            possible_cis.at[i, 'expression'] = 'high'\n",
    "            \n",
    "## ISOLATING THE VALUES WITH STRICTLY HIGH AND LOW EXPRESSIONS\n",
    "possible_cis_no_neither = possible_cis.loc[possible_cis['expression'].isin(['high', 'low'])]\n",
    "\n",
    "## GRAPH OF DEGS BESIDE POSSIBLE CIS MUTATIONS\n",
    "ax = sns.barplot(data = frequency_table, x = 'index', y = 'add_counts')\n",
    "ax.set_title('DEGs beside possible cis mutations')\n",
    "plt.xlabel('bin width from mutations')\n",
    "plt.ylabel('Number of DEGs')\n",
    "plt.savefig('/research/projects/chlamydomonas/MAexpression/analysis/DEGs/images/DEGs_near_cis_mutations.pdf', format = 'pdf', dpi = 150, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "activated-survival",
   "metadata": {},
   "outputs": [],
   "source": [
    "## GRAPH OF THE TYPE OF EXPRESSION PRODUCED BY CIS MUTATIONS\n",
    "ax = sns.countplot(x = 'bin_width', hue = 'expression', data = possible_cis)\n",
    "ax.xaxis.set_major_locator(MaxNLocator(integer=True)) \n",
    "for p in ax.patches:\n",
    "    txt = str(p.get_height().round(2))\n",
    "    txt_x = p.get_x() \n",
    "    txt_y = p.get_height()\n",
    "    ax.text(txt_x,txt_y,txt)\n",
    "ax.set_title('Type of expression produced by cis mutations')\n",
    "plt.savefig('/research/projects/chlamydomonas/MAexpression/analysis/DEGs/images/expression_near_cis_mutations.pdf', format = 'pdf', dpi = 150, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "explicit-commander",
   "metadata": {},
   "outputs": [],
   "source": [
    "## GRAPH OF THE TYPE OF EXPRESSION (WITHOUT NEITHER) PRODUCED BY CIS MUTATIONS\n",
    "a = sns.countplot(x = 'bin_width', hue = 'expression', data = possible_cis_no_neither)\n",
    "a.set_title('Type of expression produced by cis mutations')\n",
    "plt.savefig('/research/projects/chlamydomonas/MAexpression/analysis/DEGs/images/expression_without_neither_near_cis_mutations.pdf', format = 'pdf', dpi = 150, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sporting-spyware",
   "metadata": {},
   "source": [
    "### Divergence Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "vietnamese-animation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>SFS0</th>\n",
       "      <th>gene</th>\n",
       "      <th>diffs0</th>\n",
       "      <th>k4</th>\n",
       "      <th>Cincerta_transcript_ID</th>\n",
       "      <th>pi4</th>\n",
       "      <th>SFS4</th>\n",
       "      <th>aligned_sites</th>\n",
       "      <th>sites4</th>\n",
       "      <th>...</th>\n",
       "      <th>BetweennessCentralitychlamynetUnDirected</th>\n",
       "      <th>OutdegreechlamynetUnDirected</th>\n",
       "      <th>IndegreechlamynetDirected</th>\n",
       "      <th>maxDegree</th>\n",
       "      <th>maxBC</th>\n",
       "      <th>meanBC</th>\n",
       "      <th>sum_product_outdegrees</th>\n",
       "      <th>mean_product_outdegrees</th>\n",
       "      <th>meanDegree</th>\n",
       "      <th>transcript55</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g18373.t1</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>g18373</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>94.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cre08.g363350.t1.3</td>\n",
       "      <td>[399, 9, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>Cre08.g363350</td>\n",
       "      <td>146.0</td>\n",
       "      <td>0.516024</td>\n",
       "      <td>g17188.t1</td>\n",
       "      <td>0.023109</td>\n",
       "      <td>[126, 9, 0, 1, 1, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0...</td>\n",
       "      <td>1242.0</td>\n",
       "      <td>260.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cre08.g379050.t1.3</td>\n",
       "      <td>[921, 24, 13, 6, 6, 6, 1, 1, 8, 2, 0, 0, 0, 0,...</td>\n",
       "      <td>Cre08.g379050</td>\n",
       "      <td>179.0</td>\n",
       "      <td>0.497896</td>\n",
       "      <td>g13593.t1</td>\n",
       "      <td>0.053507</td>\n",
       "      <td>[273, 12, 11, 3, 5, 5, 7, 2, 6, 3, 0, 0, 0, 0,...</td>\n",
       "      <td>1882.0</td>\n",
       "      <td>415.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g8540.t1</td>\n",
       "      <td>[2623, 27, 6, 5, 0, 3, 2, 2, 3, 0, 0, 0, 0, 0,...</td>\n",
       "      <td>g8540</td>\n",
       "      <td>1414.0</td>\n",
       "      <td>0.785356</td>\n",
       "      <td>g11707.t1</td>\n",
       "      <td>0.030907</td>\n",
       "      <td>[791, 43, 11, 8, 8, 5, 8, 5, 12, 1, 0, 0, 0, 0...</td>\n",
       "      <td>10880.0</td>\n",
       "      <td>2537.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cre08.g372000.t1.2</td>\n",
       "      <td>[478, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>Cre08.g372000</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.183940</td>\n",
       "      <td>g11555.t1</td>\n",
       "      <td>0.037368</td>\n",
       "      <td>[127, 5, 3, 1, 1, 6, 0, 1, 1, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>754.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000275</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           transcript                                               SFS0  \\\n",
       "0           g18373.t1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "1  Cre08.g363350.t1.3  [399, 9, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0...   \n",
       "2  Cre08.g379050.t1.3  [921, 24, 13, 6, 6, 6, 1, 1, 8, 2, 0, 0, 0, 0,...   \n",
       "3            g8540.t1  [2623, 27, 6, 5, 0, 3, 2, 2, 3, 0, 0, 0, 0, 0,...   \n",
       "4  Cre08.g372000.t1.2  [478, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "\n",
       "            gene  diffs0        k4 Cincerta_transcript_ID       pi4  \\\n",
       "0         g18373     NaN       NaN                    NaN       NaN   \n",
       "1  Cre08.g363350   146.0  0.516024              g17188.t1  0.023109   \n",
       "2  Cre08.g379050   179.0  0.497896              g13593.t1  0.053507   \n",
       "3          g8540  1414.0  0.785356              g11707.t1  0.030907   \n",
       "4  Cre08.g372000    21.0  0.183940              g11555.t1  0.037368   \n",
       "\n",
       "                                                SFS4  aligned_sites  sites4  \\\n",
       "0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...            NaN     NaN   \n",
       "1  [126, 9, 0, 1, 1, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0...         1242.0   260.0   \n",
       "2  [273, 12, 11, 3, 5, 5, 7, 2, 6, 3, 0, 0, 0, 0,...         1882.0   415.0   \n",
       "3  [791, 43, 11, 8, 8, 5, 8, 5, 12, 1, 0, 0, 0, 0...        10880.0  2537.0   \n",
       "4  [127, 5, 3, 1, 1, 6, 0, 1, 1, 0, 0, 0, 0, 0, 0...          754.0   141.0   \n",
       "\n",
       "   ...  BetweennessCentralitychlamynetUnDirected  \\\n",
       "0  ...                                  0.000045   \n",
       "1  ...                                       NaN   \n",
       "2  ...                                  0.000001   \n",
       "3  ...                                       NaN   \n",
       "4  ...                                  0.000275   \n",
       "\n",
       "   OutdegreechlamynetUnDirected  IndegreechlamynetDirected  maxDegree  maxBC  \\\n",
       "0                          94.0                       12.0        NaN    NaN   \n",
       "1                           NaN                        NaN        NaN    NaN   \n",
       "2                           0.0                        4.0        NaN    NaN   \n",
       "3                           NaN                        NaN        NaN    NaN   \n",
       "4                           8.0                        7.0        NaN    NaN   \n",
       "\n",
       "   meanBC  sum_product_outdegrees  mean_product_outdegrees  meanDegree  \\\n",
       "0     NaN                     NaN                      NaN         NaN   \n",
       "1     NaN                     NaN                      NaN         NaN   \n",
       "2     NaN                     NaN                      NaN         NaN   \n",
       "3     NaN                     NaN                      NaN         NaN   \n",
       "4     NaN                     NaN                      NaN         NaN   \n",
       "\n",
       "   transcript55  \n",
       "0           NaN  \n",
       "1           NaN  \n",
       "2           NaN  \n",
       "3           NaN  \n",
       "4           NaN  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "divergence = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/divergence_info/divergence.out.txt', delimiter = '\\t')\n",
    "network_stat = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/divergence_info/network_divergence_diversity_statsmerged.txt', delimiter = '\\t')\n",
    "network_stat.head()\n",
    "# divergence.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "empty-straight",
   "metadata": {},
   "source": [
    "### Seaborn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "yellow-bangladesh",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "      <th>labels</th>\n",
       "      <th>genotype</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samples</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CC2344-L1</th>\n",
       "      <td>FC.per.mut</td>\n",
       "      <td>85.237884</td>\n",
       "      <td>L1</td>\n",
       "      <td>CC2344</td>\n",
       "      <td>396.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CC2344-L2</th>\n",
       "      <td>FC.per.mut</td>\n",
       "      <td>309.611696</td>\n",
       "      <td>L2</td>\n",
       "      <td>CC2344</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CC2344-L3</th>\n",
       "      <td>FC.per.mut</td>\n",
       "      <td>539.342803</td>\n",
       "      <td>L3</td>\n",
       "      <td>CC2344</td>\n",
       "      <td>63.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CC2344-L4</th>\n",
       "      <td>FC.per.mut</td>\n",
       "      <td>1271.601493</td>\n",
       "      <td>L4</td>\n",
       "      <td>CC2344</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CC2344-L5</th>\n",
       "      <td>FC.per.mut</td>\n",
       "      <td>389.713818</td>\n",
       "      <td>L5</td>\n",
       "      <td>CC2344</td>\n",
       "      <td>68.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             variable        value labels genotype  mutations\n",
       "samples                                                      \n",
       "CC2344-L1  FC.per.mut    85.237884     L1   CC2344      396.0\n",
       "CC2344-L2  FC.per.mut   309.611696     L2   CC2344       80.0\n",
       "CC2344-L3  FC.per.mut   539.342803     L3   CC2344       63.0\n",
       "CC2344-L4  FC.per.mut  1271.601493     L4   CC2344       24.0\n",
       "CC2344-L5  FC.per.mut   389.713818     L5   CC2344       68.0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAbYAAAFuCAYAAADtZD0yAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAwq0lEQVR4nO3deXxddZ3/8de7SZqmaZt0p3SnlJ1SaK1sOgouVXYFqcqAAyPqsKmj/gCdAbcZdFTccRCURRQZRKkIKIKICBTKYhcoUNrSFkr3fU/y+f1xTmiSpultc5fk5P18PO4j937u+Z7zubdtPv2e8z3fryICMzOzrOhW6gTMzMzyyYXNzMwyxYXNzMwyxYXNzMwyxYXNzMwypbzUCZTC5MmT4/777y91GmZmrVGpE+jsumSPbcWKFaVOwczMCqRLFjYzM8suFzYzM8sUFzYzM8sUFzYzM8sUFzYzM8sUFzYzM8sUFzYzM8uUohQ2SWWSnpV0T/r6akmvSXoufby/ybZXSJor6UVJ720SnyBpZvre9yUpjVdK+nUanyZpVDE+k5mZdUzF6rFdBrzQInZtRIxPH/cCSDoEmAIcCkwGfiypLN3+OuBCYGz6mJzGLwBWR8T+wLXANwr6SczMrEMreGGTNAw4Cbghh81PA26PiK0RMR+YC0ySNAToExGPR7Iy6i3A6U3a3Jw+vxM4sbE3Z2ZmXU8xemzfBb4ANLSIXyxphqSfSeqbxoYCi5pssziNDU2ft4w3axMRdcBaoH/LJCRdKGm6pOnLly9v3yfqLJa/BH/9H/jFmfDMrbBuSakzMjMruIIWNkknA8si4ukWb10HjAHGA0uAbzc2aWU30Ua8rTbNAxHXR8TEiJg4cODAHLLv5NYuhl+eDX/5Gsx9AKZeDI9+B+q2lTozM7OCKnSP7TjgVEkLgNuBEyT9IiKWRkR9RDQAPwUmpdsvBoY3aT8MeD2ND2sl3qyNpHKgBlhVmI/TiSybA6vnNY9NvxFWLyhJOmZmxVLQwhYRV0TEsIgYRTIo5KGIOCe9ZtboDGBW+nwqMCUd6TiaZJDIkxGxBFgv6ej0+tm5wN1N2pyXPj8zPcZOPbaup5WvIKL1uJlZhpRqPbZvShpP8lt2AfAJgIiYLekO4HmgDrgoIurTNp8CbgKqgPvSB8CNwK2S5pL01KYU5yN0cAMPgtqRsObVHbEJ/wJ9R5UsJTOzYlBX7NxMnDgxpk+fXuo0Cm/ZHJh1Fyx6HA79ABzwHugzdPftzKyUPKq7nbrkCtpdxqCD4IQrS52FmVlReUotMzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLFBc2MzPLlKIUNkllkp6VdE/6up+kByS9nP7s22TbKyTNlfSipPc2iU+QNDN97/uSlMYrJf06jU+TNKoYn8nMzDqmYvXYLgNeaPL6cuDBiBgLPJi+RtIhwBTgUGAy8GNJZWmb64ALgbHpY3IavwBYHRH7A9cC3yjsRzEzs46s4IVN0jDgJOCGJuHTgJvT5zcDpzeJ3x4RWyNiPjAXmCRpCNAnIh6PiABuadGmcV93Aic29ubMzKzrKUaP7bvAF4CGJrHBEbEEIP05KI0PBRY12W5xGhuaPm8Zb9YmIuqAtUD/lklIulDSdEnTly9f3s6PZGZmHVVBC5ukk4FlEfF0rk1aiUUb8bbaNA9EXB8REyNi4sCBA3NMx8zMOpvyAu//OOBUSe8HegB9JP0CWCppSEQsSU8zLku3XwwMb9J+GPB6Gh/WSrxpm8WSyoEaYFWhPpCZmXVsBe2xRcQVETEsIkaRDAp5KCLOAaYC56WbnQfcnT6fCkxJRzqOJhkk8mR6unK9pKPT62fntmjTuK8z02Ps1GMzM7OuodA9tl25BrhD0gXAQuAsgIiYLekO4HmgDrgoIurTNp8CbgKqgPvSB8CNwK2S5pL01KYU60OYmVnHo67YuZk4cWJMnz691GmYmbXGo7rbyTOPmJlZpriwmZlZpriwmZlZpriwmZlZppRqVGSXUd8QLFixkXVbtjO0topBfXqUOiUzs0xzYSugzdvq+L+nF/O1e15gW30DQ2ur+Mk5R3H4sNpSp2Zmllk+FVlAc95Yz3/ePZtt9ck0ma+t2cwXfzeTtZu3lTgzM7PscmEroEWrNu0Um7F4HSs3uLCZmRWKC1sBDW7letqYgdXUVFWUIBszs67Bha2ADh7Sh0/9035vvu5VWc5/nXE4/XtVljArM7Ns8+CRAupTVcElJ4zlfYcPYc2m7QzvV8XoAb1KnZaZWaa5sBVYz8pyxrUyCnJrXT0bt9bTt2cFXvDbzCx/XNhKYObitfzo4bnMem0tJ48bwoffMoKRA6pLnZaZWSa4sBXZghUbOefGaazdvB2An/x1HgtWbOLas8dT1b2sxNmZmXV+HjxSZHOXbXizqDW6f/YbLF69860BZma251zYiqxHxc5feWV5N8rLfJ3NzCwfXNhyFJHM+fjYKyt46Y31bE9nE9lTB+zTm6NG1DaLXXriWEb28zU2M7N88DW2HP31peVcdNszbNxWT0WZ+Mpph/HBo4bSvXzProsN6t2D7085kqcWrGLeio0cOaIvR42opVs399jMzPLBhS0Hr63ezGd+/Rwbt9UDsL0++OJvZ3LEsBoO2bdmj/c3rF9PhvXrme80zcwMn4rMyYoNW1m9qfmAj4aAN9ZuKVFGZma2Ky5sORjQuzv9q7s3i3UTDKmpKlFGZma2Ky5sORha25Pvnj2e3pXJmdvK8m5844PjGDPI02OZmXU0vsaWo7cdMJA/XHo8S9ZtoV/P7uw3sBdlHvBhZtbhuLDtgRH9qxnR38Pyzcw6Mp+KNDOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTHFhMzOzTMmpsEk6QNKDkmalr8dJ+lJhUzMzM9tzufbYfgpcAWwHiIgZwJRCJWVmZra3ci1sPSPiyRaxunwnY2Zm1l65FrYVksYAASDpTGBJwbIyMzPbS+U5bncRcD1wkKTXgPnAOQXLyszMbC/lVNgiYh7wLknVQLeIWF/YtMzMzPZOToVN0mdbvAZYCzwdEc/lPy0zM7O9k+s1tonAJ4Gh6eNC4B3ATyV9oTCpmZmZ7blcr7H1B46KiA0Akq4C7gTeDjwNfLMw6ZmZme2ZXHtsI4BtTV5vB0ZGxGZga96zMjMz20u59th+CTwh6e709SnAr9LBJM8XJDMzM7O9kOuoyK9Kug84DhDwyYiYnr790UIlZ2Zmtqdy7bEBPAu83thG0oiIWFiQrMzMzPZSrsP9LwGuApYC9SS9tgDGFS41MzOzPZdrj+0y4MCIWFnIZMzMzNor11GRi0huyDYzM+vQcu2xzQMelvQHmgzvj4jvFCQrMzOzvZRrYVuYPrqnDzMzsw4p1+H+X96bnUvqATwCVKbHujMirpJ0NfBxYHm66ZURcW/a5grgApJBKpdGxB/T+ATgJqAKuBe4LCJCUiVwCzABWAmcHREL9iZfMzPr/HIdFTkQ+AJwKNCjMR4RJ+ym6VbghIjYIKkCeDS9Hw7g2oj4VovjHEKyMvehwL7AnyUdEBH1wHUkc1Q+QVLYJgP3kRTB1RGxv6QpwDeAs3P5XGZmlj25Dh65DZgDjAa+DCwAntpdo0hsSF9WpI9oo8lpwO0RsTUi5gNzgUmShgB9IuLxiAiSHtrpTdrcnD6/EzhR6fIDZmbW9eRa2PpHxI3A9oj4a0ScDxydS0NJZZKeA5YBD0TEtPStiyXNkPQzSX3T2FCSEZiNFrNjRYHFrcSbtYmIOpLRm/1byeNCSdMlTV++fHnLt83MLCNyLWzb059LJJ0k6UhgWC4NI6I+Isan20+SdBjJacUxwHhgCfDtdPPWelrRRrytNi3zuD4iJkbExIEDB+aSupmZdUK5FravSaoB/h34HHAD8Jk9OVBErAEeBiZHxNK04DUAPwUmpZstBoY3aTaMZBqvxTQvpI3xZm0klQM1wKo9yc3MzLIjp8IWEfdExNqImBUR74yICRExdXftJA2UVJs+rwLeBcxJr5k1OgOYlT6fCkyRVClpNDAWeDIilgDrJR2dXj87F7i7SZvz0udnAg+l1+HMzKwL2pNRkR8HRjVtk15ra8sQ4GZJZSRF9I6IuEfSrZLGk5wyXAB8It3fbEl3kCyFUwdclI6IBPgUO4b735c+AG4EbpU0l6SnNiWXz2RmZtmkXDo3kh4D/kayWnZjoSEiflO41Apn4sSJMX369N1vaGZWfB7V3U65zjzSMyL+X0EzMTMzy4NcB4/cI+n9Bc3EzMwsD9rssUlaz47h9ldK2koy9F8k91/3KXyKZmZmuWuzsEVE72IlYmZmlg+5joo8g2QY/dr0dS3wjoj4XeFS65rWbd7OzNfWMn/FBobUVHH4sBoG9e6x+4ZmZgbkPnjkqoj4beOLiFgj6SrgdwXJqouqq2/gl9MWcs39c96MnXLEvnz19EOprfJqQWZmuch18Ehr2+VaFC1Hr67cxHceeKlZ7Pf/eJ2Xl27YRQszM2sp1+I0XdJ3gB+RDCa5hOSeNsujzdvr2VbfsFN8w5a6EmRjZntjzpJ1PDp3BWs3b+dtYwdwxLBaKivKSp1Wl5Jrj+0SYBvwa+AOYDNwUaGS6qqG9a3iiOE1zWJ9qsrZb0B1iTIysz0x5411fOj6x/naH17gBw/N5ezrn+DxeStLnVaXk+tckRsj4vLG2fEj4sqI2Nj4vqQfFC7FrqO2Z3e+ddYRnHLEEHp2L+Po0f245fxJjHRhM+sUnpi3knWbd5xhiYDv/fllNm71WZdiytd1suPytJ8ub+yg3nzrrCNYtWEbfXpUUN3DlzLNOouNW+t3iq3bsp36huzMyy5pFHBsRPyy1LnsSq6nIq2IKsvLGFJb5aJm1skcvV8/urWY6fHjb9uPPlUVpUmoMEYBHyl1Em1xYTMzy5Nxw2q59YK3csyY/hw4uDffPHMc7zl0cNHzkPQfkuZIekDSryR9TtIYSfdLelrS3yQdlG57k6TvS3pM0jxJZ6ZxSfofSbMkzZR0drr7a4C3SXpO0mfSfY1vcuy/Sxon6ep0JZeHJL0s6eNNtvm8pKckzZD05Xx//nx1CTwbtZl1eRVl3Thu/wFMGNmX+oYGqiuL31OTNBH4IHAkye/4Z0hGsV8PfDIiXpb0VuDHwAlpsyHA8cBBJGtc3gl8ABgPHAEMAJ6S9AhwOfC5iDg5Pd4q4GPApyUdAFRGxAxJHwDGAUcD1cCzkv4AHEay1uYkktoxVdLbI+KRfH0H+Sps38vTfszMOr0eFWVAyYb4Hw/cHRGbAST9HugBHAv8X7JWMwCVTdr8LiIagOclDW6yn1+la2IulfRX4C3AuhbH+z/gPyR9HjifZN3MRo15bJb0F5JidjzwHuDZdJteJIWuOIUt/UJ2edUzIk5Nf96Ur4TMzKxdWjuD1g1YExHjd9FmayvtczoTFxGbJD0AnAZ8CJjY9O2Wm6f7/e+I+N9c9r83dneN7VvAt4H5JPeu/TR9bABmFSopMzPba48Cp0jqIakXcBKwCZgv6Sx48/rZEbvZzyPA2ZLKJA0E3g48CawHWk6QfwPwfeCpiFjVJH5amkd/4B3AU8AfgfPT3JA0VNKgdnzenexudv+/pgf+akS8vclbv0/PtZqZWQcSEU9Jmgr8A3gVmA6sBT4KXCfpS0AFcHu6za78Fjgm3SaAL0TEG5JWAnWS/gHcFBHXRsTTktYBP2+xjyeBPwAjgK9GxOvA65IOBh5PT4tuAM4BluXh4wOgiN3fXyHpBeCkiJiXvh4N3BsRB+crkWKaOHFiTJ8+vdRpmJm1pt2D8ST1iogNknqS9LwujIhn2p/aLo+3L/AwcFB6rQ5JVwMbIuJbhTruruQ6eOQzwMOS5qWvRwGfKEhGZmbWXtdLOoRk0MjNBS5q5wJfBz7bWNRKLaceG4CkSpKhoABzImJrW9t3ZO6xmVkH5tun2ml3oyI/sIu3xkgiIu4qQE5mZmZ7bXenIk9p470AXNjMzKxD2d2oyH8pViJmZmb5kNNckZJqJH1H0vT08W1JNbtvaWZmVly5ToL8M5Kb8j6UPlq7X8HMzDJE0j6Sbpf0iqTnJd0r6YD0ca+kuZJekHSHpMGS3p1Osjwz/XlCk33dL+kfkmZL+omkshbHOlNSpHNdNo33kfSapB/mmneuw/3HRMQHm7z+sqTncj2ImZl1Lkrunv4tye0CU9LYeGAwSWfnsxHx+zT+TmAgsAI4JSJel3QYySwjQ9Ndfigi1qX7vRM4i+QmcST1Bi4FprWSyleBv+5J7rkWts2Sjo+IR9MkjiOZYsvMzEps1OV/+AjwXyQzfCwErlxwzUntXQj0ncD2iPhJYyAinpN0PvB4Y1FL439ppf1soIekyojYGhGNkyeXA91pPo/kV4FvAp9rugNJE0gK6f00n4OyTbmeivwk8CNJCyQtAH6Ib9A2Myu5tKj9FBhJcg/cSOCnabw9DiNZ7ibXeEsfBJ5tes+zpD+STJ21nqTXhqQjgeERcU/TxpK6kcxV/Pk9TbzNwibpsvRpr4g4gmRtnXERcWREzNjTg5mZWd79F9CzRaxnGi8JSYcC36BFBygi3kuy9lslcEJavK4F/r2V3fwbydSNi/b0+LvrsTUO9/9BmtS6Jt1JMzMrvRF7GM/VbGDCHsQBkDSM5NrcuRHxSsv3I2ILyWKmp5GsEnAYyZSNC0gWJZ2aDiA5Brg4jX8LOFfSNbkkvrvC9kK60wPTJbwbHzMlucdmZlZ6C/cwnquHgEpJH28MSHoLMBc4VtJJTeKTJR0uqZZkNv8rIuLvTd7vJWlI+rwceD/J1IxrI2JARIyKiFHAE8CpETE9Ij4aESPS+OeAWyLi8lwS390N2h+WtA/JyJZTc9mhmZkV1ZUk19iano7clMb3WkSEpDOA70q6HNgCLAA+DZycxr8LbAdmAJcBFwP7k6yo/R/prt5Dcu1vajrncBlJ0XxzUEq+5TwJcps7kX7T4naADs2TIJtZB7bHkyAXaFRkp5WvwvZsRByZh3yKwoXNzDowz+7fTrkO99+d9ldHy9mmbXW8tnoTG7ZsL3UqZmYdTq43aFsH8cKSdVxz3xwee2UFR43oyxdPOphxw2pLnZaZWYeRrx6bu85FsHz9Fj75i6f560vL2V4fTJu/ivNveorXVnsSGDOzRvkqbP8vT/uxNixctZlXV25qFluxYRsLVm4sUUZmZh3P7lbQnknr189EMhp0HMmTPxUgN2uhurKMsm6ivqH5H0nvHj6jbGbWaHe/EU8uShYZsHz9Fp55dTWzXlvHQUP6MHFkLYNrqvJ6jP0GVHPxO/fnew++/GbsvGNGMmZQr7wex8wMkmVrgO8CbwG2suM+NtL4AST3sc0ELiGZdvEakkmOtwGfj4iH0n2dDXyR5D62P0TEF9L4J4GLgHpgA3BhRDyfvnc/yWwkj0ZEzvUoL8P9O5t8D/ffvK2Or/3hBW6btuNG/1PGDeHrZxxOn6qKvB0HYO3mbcx6bR2LV29iSE0Vhw3tQ7/qyrwew8xKqkOMWUiXl3mMZNman6Sx8STTYLW2bM1yoAJY2nTZmogYKqk/8CwwISKWS7qZZCaRByX1aZyqUdKpwL9FxOT09YkkN55/Yk8KW07nsCQdTTJf5MEklbgM2BgRfXI9UJbNW7GxWVED+P2MJZx//GiOHNE3r8eqqerOcfsPyOs+zayTu7pmpxu0uXpth1m2BtgPeCkilqfv/Zlk9v8HW8w/XE2Ty19p4XvHniae6+CRHwIfBl4GqoB/JZ0Y2WBrXUOr8S3bW4+bmeVNUtR2WrYmjbdHPpetmQscJGlUOlfk6cDwxg0lXSTpFZI12S5tZ965j4qMiLlAWUTUR8TPSaq5AaP69+SQIc07ryP6VbHfgOoSZWRmXUiHX7YmIlYDnwJ+DfyN5FpdXeP2EfGjiBhDMsL+S+09fq7D6TZJ6g48J+mbwBKSLqMB/aor+f6Hx3PTYwv4y5zlHLd/fy44fj8G1/QodWpmln2FXLbmzF3E/2lXjXa1bE166rLxmtyFJINFWroduK4dOQO599j+Od32YmAjSRfyA+09eJbsP6g3V59yKL+/5Hi+fvrhHLhP71KnZGZdQ4dftibdZlD6sy/JIqI3pK/HNtnsJJJLXu2Sa2E7PSK2pAuNfjkiPotvBdhJeVk3+lV3p6I8X/e9Z9D2zTDvYbjr4/D7T8OiadDQ2n/czCxHV5IsU9NUXpatAc4A3i3pFUmzgauB10l+/18i6WVJzwMfA5bRfNma59LHoHSX30u3/TtwTUS8lMYvljRb0nPAZ4HzGnOQ9Dfg/4ATJS2W9N5ccs9puL+kZyLiqBaxTjWjf1N7Oty/oSHYvL2e6krfCN1uL/8ZbmuywlG3cviX+2D4pNLlZNax7Plw/8KMiuy0djfzyIeBjwCjJU1t8lYfYGUhE+soXnxjHbc9sZBp81fxvsP34YwjhzKyvy8v7pX6bfB4i8G0DXXwwj0ubGbtkRSxLlvIWtpdF+QxkoEiA4BvN4mvJ1kxNdOWrNmcTDK8ZgsALy5dz4xFa/j+h4+kV4/83njdZTS0dgtE15skwMwKp82LQRHxakQ8HBHHAHNI7jjvDSyOiLq22mbB3OUb3ixqjR56cTmvrmp5OttyUtYdjr24eaxbGRzky7Vmlj85jXKQdBbwJHAW8CFgmqTWhoFmSkW3nb+eboLybh1ixpvOadTb4KN3woEnweEfgvPugaETSp2VmWVIrqMhvgS8JSKWAUgaSDIlyp2FSqwjGDu4FxNH9mX6q6vfjP3z0SMZ5Ruv9173njD23cnDzKwAci1s3RqLWmol+VvLrcPq36uSa88+gsdeWcmMxWs5Zr/+TNqvH5XlZaVOzczMdiHXwnafpD8Cv0pfnw3cu7tGknoAjwCV6bHujIirJPUjmVplFMnUKh9Kp1xB0hXABSR3pV8aEX9M4xOAm0jmqrwXuCwiIp1g8xZgAknBPTsiFuT4uXZreL9qzu5XzdlvydcezcyskHLtdQXwvyRr7RwBXJ9ju63ACRFxBDAemJyuFHA5yazOY4EH09dIOgSYAhwKTAZ+LKmxe3QdcCEwNn1MTuMXAKsjYn/gWpL5yczMrIvKtbC9OyLuiojPRsRnIuK3wPt21ygSG9KXFekjgNOAm9P4zSQzPZPGb4+IrRExn2TqlkmShgB9IuLx9G74W1q0adzXnSR3qHt0h5lZF9VmYZP0KUkzgQMlzWjymE+O97FJKkunSlkGPBAR04DBEbEEIP3ZOOXKUGBRk+aL09jQ9HnLeLM26S0Ia4H+ueRmZmbZs7trbL8E7gP+m/R0YWp9RKzK5QARUQ+MTyfH/G26ququtNbTijbibbVpvuNkNukLAUaMaO+k12Zm1lG1WdgiYi1JD+jD7T1QRKyR9DDJtbGlkoZExJL0NGPjiMvFNFl8DhhGMuHm4vR5y3jTNovTBexqgJ2KbkRcT3ptcOLEiZ7qwswsowo6ZF/SwLSnhqQq4F0kM5hMZccMzucBd6fPpwJTJFVKGk0ySOTJ9HTleklHp9fPzm3RpnFfZwIPRS4zO5uZWSYVerr6IcDN6cjGbsAdEXGPpMeBOyRdQDIT9VkAETFb0h3A8ySrq16UnsqEZPXVm0iG+9+XPgBuBG6VNJekpzalwJ/JzMw6sJyWrcmaPV22xsysiDyqu50yP3uImZl1LS5sZmaWKS5sZmaWKS5sWbR5LdRtLXUWZmYlUehRkVZo27fAqldg+2boOQBevA+e/hn03Q/e9hkY/lbwDGNm1oW4sHVmG5bD374NT/4EImCfcXDIqbDipeQx/2H41wdhn7YmezEzyxafiuzMFj8F065LihrAGzNg2Qsw+NDkdd0WWDqrdPmZmZWAC1tn9kYr81AvfCLpuTUqryxePmZmHYALW2c28KCdY0PGwYoXk+d9hsGQ8UVNycys1HyNrTMbPgkOOQ2eT6fNrBkOR/8bvPIQjP8ojH479Btd2hzNzIrMU2p1dpvXJANFtm+G/vtDzdDdNjGzDs3DmNvJPbbOrqo26bmZmRnga2xmZpYxLmxmZpYpLmxmZpYpvsbWQSxZu5nXVm+md48K9htQTUW5/89hZrY3XNg6gH8sWsPHb5nOsvVbKe8mPj/5QM5560iqK/3HY2a2p9wtKLG1m7Zx5W9nsmx9Mht/XUPw3/fO4YUl60qcmZlZ5+TCVmIrN25j9us7F7HFqzeXIBszs87Pha3E+vbszoH79N4pvm9tjxJkY2bW+bmwlVjf6u58/fTDqKmqAJKl0y49cX8O3qdPiTMzM+ucPDqhA5g4qh/3XHI8C1dtoraqgjGDqulR4T8aM7O94d+eHcTwfj0Z3q9nqdMwM+v0fCrSzMwyxYXNzMwyxYXNzMwyxYXNzMwyxYNHsmjdG7BpBVQPgt6DSp2NmVlRubBlzfxH4LefgHWvQ+0IOON/YeSxpc7KzKxofCoyS1bOg9s/mhQ1gDUL4dfnwJpFpc3LzKyIXNiyZO0i2Npi3slNK5O4mVkX4cKWJdUDQC3+SMu6Q89+pcnHzKwEXNiypP9YePdXm8cmXwP99i9NPmZmJeDBI1lS3h0mng8jjkmus9UMhUEHQ5n/mM2s6/BvvKzp3hOGTQAmlDoTM7OScGHryuq2wYqXYctqqB0JtcNLnZGZWbu5sHUC67dsZ+v2Bgb0rszfTrdugKdugIe+Ag31ycCTKb+C4ZPydwwzsxLw4JEOrK6+gUdfXs45N0zjpB/8jR8+9DJL1m7Oz86XzoI/X5UUNYCNK2DqpbBxZX72b2ZWIu6xdWCzXl/LuT97koZIXn/rTy9R1xBcduJYJLVv52sX7xxb/kJy31t1//bt28yshNxj68Bmv77uzaLW6ObHFrB8/db277xm2M6xAQdCTxc1M+vcXNg6sF6VO3eo+/XqTmV5Hv7YBh8KJ16144bunv3htB+4t2ZmnZ5PRXZgRwyrZVjfKhavTq6rSXDF5IOp6dm9/Tuv7A1HXwQHvBc2r04mTK4d0f79mpmVmAtbBzZqQDW3nj+JZxauYfWmbYwfXsu4YTX5O0BFZdJzMzPLEBe2Dm70wF6MHtir1GmYmXUavsZmZmaZ4sJmZmaZ4lORXdGGZbB0NmzfAgMPgP5jSp2RmVneuLB1NWsWwl0XwsLHk9c9auGffwdDjyxlVmZmeeNTkV3Nwmk7ihrAljXwt28lvTczswxwYetqVs3bOfb6s7BtQ/FzMTMrABe2rqa1U46HngFV/Yqfi1mutm1KVqQwy4ELW1czbBK862oo75G8PuD9MPEC6Oa/CtYBbd8CL/0Jbj0dfjYZZtwBW9aWOivr4BQRu98qYyZOnBjTp08vdRql09AAq+dD/bZkGq3u1aXOyKx18/8GN5/cPHbmTXDYGSVJp0jauXSH+b/pXVG3bskQ/0EHu6hZxzbnnp1j066DujyscGGZ5cJmZh1Xj9qdY1X98K8ua4v/dljbNq+Dlx+AP30Jpv8MVs4tdUbWlRz4vuZnFbqVwbEXQ3lF6XKyDq+gN2hLGg7cAuwDNADXR8T3JF0NfBxYnm56ZUTcm7a5ArgAqAcujYg/pvEJwE1AFXAvcFlEhKTK9BgTgJXA2RGxoJCfq0uZ+Wu493M7XvcbA+feDbXDS5eTdR37jod/uQ/mPZwMJBnzTtj3qFJnZR1coWceqQP+PSKekdQbeFrSA+l710bEt5puLOkQYApwKLAv8GdJB0REPXAdcCHwBElhmwzcR1IEV0fE/pKmAN8Azi7w5yqN156BOffCtvVw8MkwdFKy9EyhrF0ED36leWzVK/DGTBc2K54hRyQPsxwVtLBFxBJgSfp8vaQXgKFtNDkNuD0itgLzJc0FJklaAPSJiMcBJN0CnE5S2E4Drk7b3wn8UJIia8M9X38Wfv4+qEtnCJn2EzjnLtj/xMIds74O6jbvHPeFezPrwIp2jU3SKOBIYFoauljSDEk/k9Q3jQ0FFjVptjiNDU2ft4w3axMRdcBaoH8rx79Q0nRJ05cvX97y7Y7v5Qd2FLVGj36X+m1bmLF4Db+atpDfPfsa85bn8SbWmmEw4YLmse69YPDB+TuGmVmeFWUSZEm9gN8An46IdZKuA74KRPrz28D5tH7/RrQRZzfv7QhEXA9cD8l9bHv6GfJq7WvJab6qfsmw+25lu2+zvZWe0/aNzF26jg9cN526huQjDe5TyW3/+lb2H9S7/XmWVcBxl0CffeG5X8CgQ+DYS2HgQe3ft5lZgRS8sEmqIClqt0XEXQARsbTJ+z8FGm9WWQw0vXgzDHg9jQ9rJd60zWJJ5UANsCr/nyRPFk2DX5+TLB1TXgnv+yaMOxsqqtpud8B74e/fhWh4M9RwzKV8/YH5bxY1gKXrtjJt3qr8FDZIem3HXwYTPpbkWN49P/s1MyuQgp6KlCTgRuCFiPhOk/iQJpudAcxKn08FpkiqlDQaGAs8mV6rWy/p6HSf5wJ3N2lzXvr8TOChDnt9bcNyuOsTSVGD5FrV7y+DZXN233boBDh3Kox9L4w4Fj50C5uHv52FKzfttOnyDQW4BlZV46JmZp1CoXtsxwH/DMyU9FwauxL4sKTxJKcMFwCfAIiI2ZLuAJ4nGVF5UToiEuBT7Bjuf1/6gKRw3poONFlFMqqyY9q4LJnKqqW1i3a/HlpZBYx+G4w4Jum1lXenGjj3mFF85Z7nm2361tE7XWI0M+syCj0q8lFavwZ2bxttvg58vZX4dOCwVuJbgLPakWbxVPVPrlete715vPeQ1rdvTVnzP7KTxg1hS109N/5tPn2qyvnC5IM4ckRt+3M1M+ukPAlysc37K9z+kWT9M3WDE6+Ct3wcKts3Z+Oy9Vuo6NaNvtU+XWjWyXkS5HYqyqhIa2K/f4JPPAJrFkLPATBgLFT0aPduB/Vu/z7MzLLAha0U+o9JHmZmlneeBNnMzDLFhc3MzDLFpyJt7zU0wJJn4dXHoKwSRh4L++w0cNXMrKhc2GzvLXoCbj4FGuqS15W94WP3wpBxpc3LzLo0n4q0vVNfB4//aEdRA9i6Hl76Y+lyMjPDhS3bNq2GLesKs++GOlj/xs7xxunCzMxKxIUtizatgmdugRtOgJ+/H+bc0/rqAO1R0QPe+omd4weflN/jmJntIRe2LJr7Z5h6CayaB0tnwu0fhUVP5v84+78bTv1Rck/e4MPh7Ntg+Fvzfxwzsz3gwSNZs31Lsrp2Sy/em8x6kk89+8JR58DBJ4PKoEeelsoxM2sHF7ZOaM2mbby4dD3rN9cxakDP5muvqQx6Dd65UfXAwiVUVVu4fZuZ7SEXtgKbv2IDLy3dwIYtdRywT28O2qc3FWV7fwZ4xYatfO2e5/ndc8kKAVUVZdx0/lt2LFVTXgHHXgwv/6nJMPw+yTpuZmZdgAtbAT376iquffBlHnlpBQDdBD85ZwLvOXSfvd7n7NfWvlnUADZvr+equ2fz6wuPpqZnOrP/8GPg/D/BwieSQR4jjobBh7brs5iZdRYubAWydO0Wnpi/6s2iBtAQ8KXfzWL88FoG9dm72fiXrd95dewXl65n3Za6HYWtWzcYNiF5mJl1MR4VWSBvrNvCpm31O8WXrd/Khq11rbTIzegBO6/bduJBgxjQq3Kv92lmliUubAXSr7o71d3L6NZiycC3jx3A4Jq9XzvtkH378PXTD6Nn9zIAjhhewxcmH0RV+trMrKvzCtoF9Jc5S3lp6QZufHQ+y9Zv5bgx/fnPUw7hwH36tGu/EcGrqzaxeWs9+/btQU2VV802yxCvoN1OvsZWQG8bO5BhfXsyfngtFWXd2G9gT2p7tv+UoSRG9d/5lKSZmbmwFVR5WTfGDvZNy2ZmxeRrbGZmlikubGZmlikubGZmlikubGZmlikubGZmlikubGZmlikubGZmlim+j61IVmzYyj8WreHVlZvYb2A144bV0K/a8zuameWbC1sRbNhSx7f/+CK/emrRm7F/PX40n3vvgfSo8ByPZmb55FORRTBvxYZmRQ3gxr/PZ97yjSXKyMwsu1zYimDj1p2Xr4mAzdv3fvkaMzNrnQtbEYwa0JN9a5svVTNmYDUj+nkiYzOzfHNhK4IhNVXccO5E3nPwYGqqKnj/4fvwo48excDeHjxiZpZvHjxSJIfsW8P3P3Ikazdtp7a6gspyDxoxMysEF7Yi6lFRRo8aFzQzs0LyqUgzM8sUFzYzM8sUFzYzM8sUFzYzM8sUFzYzM8sUFzYzM8sUFzYzM8sUFzYzM8sUFzYzM8sUFzYzM8sURUSpcyg6ScuBV0uYwgBgRQmP35Lz2b2OlpPzaVtnzmdFREwuZDJZ1yULW6lJmh4RE0udRyPns3sdLSfn0zbn07X5VKSZmWWKC5uZmWWKC1tpXF/qBFpwPrvX0XJyPm1zPl2Yr7GZmVmmuMdmZmaZ4sJmZmaZ4sJWBJIWSJop6TlJ09NYP0kPSHo5/dm3gMf/maRlkmY1ie3y+JKukDRX0ouS3lukfK6W9Fr6HT0n6f1FzGe4pL9IekHSbEmXpfGSfEdt5FOS70hSD0lPSvpHms+X03ipvp9d5VOyv0PpMcokPSvpnvR1yf6NdXkR4UeBH8ACYECL2DeBy9PnlwPfKODx3w4cBcza3fGBQ4B/AJXAaOAVoKwI+VwNfK6VbYuRzxDgqPR5b+Cl9Lgl+Y7ayKck3xEgoFf6vAKYBhxdwu9nV/mU7O9QepzPAr8E7klfl+zfWFd/uMdWOqcBN6fPbwZOL9SBIuIRYFWOxz8NuD0itkbEfGAuMKkI+exKMfJZEhHPpM/XAy8AQynRd9RGPrtS6HwiIjakLyvSR1C672dX+exKwf8OSRoGnATc0OK4Jfk31tW5sBVHAH+S9LSkC9PY4IhYAskvMmBQkXPa1fGHAouabLeYtn+p5tPFkmakpyobT9sUNR9Jo4AjSXoBJf+OWuQDJfqO0tNszwHLgAcioqTfzy7ygdL9Hfou8AWgoUms5H9/uioXtuI4LiKOAt4HXCTp7aVOqA1qJVaMe0KuA8YA44ElwLeLnY+kXsBvgE9HxLq2Ni1GTq3kU7LvKCLqI2I8MAyYJOmwNjYvVT4l+X4knQwsi4inc21SyHzMha0oIuL19Ocy4Lckpx2WShoCkP5cVuS0dnX8xcDwJtsNA14vdDIRsTT9ZdUA/JQdp2aKko+kCpIicltE3JWGS/YdtZZPqb+jNIc1wMPAZDrA36Gm+ZTw+zkOOFXSAuB24ARJv6ADfD9dlQtbgUmqltS78TnwHmAWMBU4L93sPODuIqe2q+NPBaZIqpQ0GhgLPFnoZBp/AaTOIPmOipKPJAE3Ai9ExHeavFWS72hX+ZTqO5I0UFJt+rwKeBcwh9J9P63mU6rvJyKuiIhhETEKmAI8FBHn0MH+jXUppR69kvUHsB/JCKh/ALOBL6bx/sCDwMvpz34FzOFXJKdmtpP8b/GCto4PfJFkpNaLwPuKlM+twExgBsk//CFFzOd4klNBM4Dn0sf7S/UdtZFPSb4jYBzwbHrcWcB/7u7vcInyKdnfoSbHeQc7RkWW7N9YV394Si0zM8sUn4o0M7NMcWEzM7NMcWEzM7NMcWEzM7NMcWEzM7NMcWEz2wVJtZL+bU+3k7SvpDsLm52Z7YqH+5vtQjpP4z0R0db0UTlvZ2bF4R6bZZqkUZLmSLpB0ixJt0l6l6S/p+tkTUrX8fpckzaz0mJ1DTAmXdvrfyT1kvSgpGeUrK93Wtqk5XajlK41l64d9vN0+2clvTONf0zSXZLuT/P4Zhovk3RTmsNMSZ8p6hdmlgHlpU7ArAj2B84CLgSeAj5CMrvHqcCVJDN7tOZy4LBIJttFUjlwRkSskzQAeELS1Fa2G9VkHxcBRMThkg4iWeXhgPS98SQz928FXpT0A5IZ4Ic29v4ap44ys9y5x2ZdwfyImBnJ5LizgQcjOQc/Exi1B/sR8F+SZgB/JllqZPBu2hxPMtUTETEHeBVoLGwPRsTaiNgCPA+MBOYB+0n6gaTJQFurDJhZK1zYrCvY2uR5Q5PXDSRnLepo/m+hxy7281FgIDAh7Z0tbWPbRq0tUdJaXvVAeUSsBo4gmbH+IpovXGlmOXBhM4MFwFEAko4CRqfx9UDvJtvVkKy7tT29VjZyF9s19QhJQSQ9BTmCZOLbVqWnOLtFxG+A/2jMy8xy52tsZsm6Z+emKzI/BbwEEBEr00Ems4D7gG8Av5c0neS63JxdbPejJvv+MfATSTNJeoYfi4ityco0rRoK/FxS4386r8jfxzTrGjzc38zMMsWnIs3MLFNc2MzMLFNc2MzMLFNc2MzMLFNc2MzMLFNc2MzMLFNc2MzMLFP+P7ZmgX8lC9vFAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 435.25x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "names = ['L1', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7', 'L8', 'L9', 'L10', 'L11', 'L12', 'L13', 'L14', 'L15']\n",
    "\n",
    "# SUM OF FOLD_CHANGE PER SAMPLE\n",
    "\n",
    "FC_stats = pd.DataFrame()\n",
    "data = pd.DataFrame()\n",
    "data['samples'] = CC2344[1:] + CC2931[1:]\n",
    "data.set_index('samples', inplace = True)\n",
    "\n",
    "for i in names:\n",
    "    for j in list(data.index.values):\n",
    "        if i in j:\n",
    "            data.at[j, 'labels'] = i\n",
    "\n",
    "for i in CC2344[1:] + CC2931[1:]:\n",
    "    data.at[i, 'generations'] = generations[i]\n",
    "    data.at[i, 'mutations'] = mutations[i]\n",
    "    if 'CC2344' in i:\n",
    "        data.at[i, 'genotype'] = 'CC2344'\n",
    "        data.at[i, 'total_fold_change'] = abs(CC2344_df[i]).sum()\n",
    "        data.at[i, 'CC2344_total_FC'] = abs(CC2344_df[i]).sum()\n",
    "        data.at[i, 'FC.per.mut'] = abs(CC2344_df[i]).sum()/mutations[i]\n",
    "        data.at[i, 'FC.per.exp.mut'] = abs(CC2344_df[i]).sum()/expected_mut.at[i, 'exp_mut']\n",
    "        data.at[i, 'FC.per.gen'] = abs(CC2344_df[i]).sum()/generations[i]\n",
    "    if 'CC2931' in i:\n",
    "        data.at[i, 'genotype'] = 'CC2931'\n",
    "        data.at[i, 'total_fold_change'] = abs(CC2931_df[i]).sum()\n",
    "        data.at[i, 'CC2931_total_FC'] = abs(CC2931_df[i]).sum()\n",
    "        data.at[i, 'FC.per.mut'] = abs(CC2931_df[i]).sum()/mutations[i]\n",
    "        data.at[i, 'FC.per.exp.mut'] = abs(CC2931_df[i]).sum()/expected_mut.at[i, 'exp_mut']\n",
    "        data.at[i, 'FC.per.gen'] = abs(CC2931_df[i]).sum()/generations[i]\n",
    "for i in CC2344[1:]:\n",
    "    temp = CC2344_df[i].mask(CC2344_df[i] > 0, 0, inplace = False)\n",
    "    data.at[i, 'CC2344_down_reg'] = temp.sum()\n",
    "    temp = CC2344_df[i].mask(CC2344_df[i] < 0, 0, inplace = False)\n",
    "    data.at[i, 'CC2344_up_reg'] = temp.sum()\n",
    "        \n",
    "for i in CC2931[1:]:\n",
    "    temp = CC2931_df[i].mask(CC2931_df[i] > 0, 0, inplace = False)\n",
    "    data.at[i, 'CC2931_down_reg'] = temp.sum()\n",
    "    temp = CC2931_df[i].mask(CC2931_df[i] < 0, 0, inplace = False)\n",
    "    data.at[i, 'CC2931_up_reg'] = temp.sum()\n",
    "    \n",
    "####################################################################################################################################################################################\n",
    "    \n",
    "# data_1 = pd.melt(data, id_vars = ['labels'], value_vars = ['CC2344_down_reg', 'CC2344_up_reg', 'CC2931_down_reg', 'CC2931_up_reg'])\n",
    "\n",
    "# for i in list(data_1.index.values):\n",
    "#     if 'CC2344' in data_1.at[i, 'variable']:\n",
    "#         data_1.at[i, 'genotypes'] = 'CC2344'\n",
    "#     if 'CC2931' in data_1.at[i, 'variable']:\n",
    "#         data_1.at[i, 'genotypes'] = 'CC2931'\n",
    "#     if 'up' in data_1.at[i, 'variable']:\n",
    "#         data_1.at[i, 'regulated'] = 'up'\n",
    "#     if 'down' in data_1.at[i, 'variable']:\n",
    "#         data_1.at[i, 'regulated'] = 'down'\n",
    "        \n",
    "#################################################################################################################################################################################### \n",
    "\n",
    "data_2 = pd.melt(data.reset_index(), id_vars = 'samples', value_vars = ['FC.per.mut', 'FC.per.exp.mut'])\n",
    "data_2.set_index('samples', inplace = True)\n",
    "\n",
    "for i in names:\n",
    "    for j in CC2344[1:] + CC2931[1:]:\n",
    "        if i in j:\n",
    "            data_2.at[j, 'labels'] = i\n",
    "for i in list(data_2.index.values):\n",
    "    if 'CC2344' in i:\n",
    "        data_2.at[i, 'genotype'] = 'CC2344'\n",
    "    if 'CC2931' in i:\n",
    "        data_2.at[i, 'genotype'] = 'CC2931'\n",
    "    data_2.at[i, 'mutations'] = mutations[i]\n",
    "\n",
    "####################################################################################################################################################################################\n",
    "\n",
    "# FC_stats['CC2344.UP'] = data['CC2344_up_reg'].describe()\n",
    "# FC_stats['CC2344.DOWN'] = data['CC2344_down_reg'].describe()\n",
    "# FC_stats['CC2931.UP'] = data['CC2931_up_reg'].describe()\n",
    "# FC_stats['CC2931.DOWN'] = data['CC2931_down_reg'].describe()\n",
    "\n",
    "# FC_stats['CC2344.TOTAL'] = data['CC2344_total_FC'].describe()\n",
    "# FC_stats['CC2931.TOTAL'] = data['CC2931_total_FC'].describe()\n",
    "\n",
    "# for i in CC2344[1:] + CC2931[1:]:\n",
    "#     if 'CC2344' in i:\n",
    "#         FC_stats[i] = CC2344_df[i].describe()\n",
    "#     if 'CC2931' in i:\n",
    "#         FC_stats[i] = CC2931_df[i].describe()   \n",
    "\n",
    "############################################################################################################################\n",
    "\n",
    "# sns.set_theme()\n",
    "# sns.relplot(data = data, kind = 'scatter', x = 'generations', y = 'total_fold_change', hue = 'genotype', size = 'mutations')\n",
    "# sns.relplot(data = data, kind = 'scatter', x = 'generations', y = 'total_fold_change', hue = 'samples', size = 'mutations')\n",
    "# sns.relplot(data = data, kind = 'scatter', x = 'generations', y = 'total_fold_change', hue = 'samples')\n",
    "# sns.relplot(data = data, kind = 'scatter', x = 'generations', y = 'total_fold_change', hue = 'genotype')\n",
    "# sns.relplot(data = data, kind = 'scatter', x = 'mutations', y = 'total_fold_change', hue = 'samples')\n",
    "# sns.relplot(data = data, kind = 'scatter', x = 'mutations', y = 'total_fold_change', hue = 'genotype')\n",
    "# sns.relplot(data = data, kind = 'scatter', x = 'generations', y = 'mutations', hue = 'genotype')\n",
    "# g = sns.relplot(data = data_1, kind = 'scatter', x = 'labels', y = 'value', hue = 'genotypes', style = 'regulated')\n",
    "# g.set(ylim = (-30000,30000))\n",
    "# g.set_axis_labels(\"MA lines\", \"fold change\")\n",
    "\n",
    "# sns.relplot(data = data, kind = 'scatter', x = 'FC.per.mut', y = 'FC.per.exp.mut', hue = 'genotype')\n",
    "sns.relplot(data = data, kind = 'scatter', x = 'mutations', y = 'total_fold_change', hue = 'genotype')\n",
    "\n",
    "############################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "administrative-manual",
   "metadata": {},
   "source": [
    "### Log2Fold Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "resident-thing",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# for i in CC2931_res:\n",
    "#     temp = CC2931_log2Fold[i].mask(CC2931_log2Fold[i] < 0)\n",
    "#     stats.at[i, 'sum_up'] = temp.sum()\n",
    "    \n",
    "#     temp = CC2931_log2Fold[i].mask(CC2931_log2Fold[i] > 0)\n",
    "#     stats.at[i, 'sum_down'] = temp.sum()\n",
    "#     stats['difference'] = stats['sum_up'] + stats['sum_down']\n",
    "#     stats['CC2931_total_dif'] = stats['difference'].iloc[[0,12]].sum()\n",
    "#     stats['CC2931_total_avg'] = stats['difference'].iloc[[0,12]].mean()\n",
    "    \n",
    "# for i in CC2344_res:\n",
    "#     temp = CC2344_log2Fold[i].mask(CC2344_log2Fold[i] < 0)\n",
    "#     stats.at[i, 'sum_up'] = temp.sum()\n",
    "    \n",
    "#     temp = CC2344_log2Fold[i].mask(CC2344_log2Fold[i] > 0)\n",
    "#     stats.at[i, 'sum_down'] = temp.sum()\n",
    "#     stats['difference'] = stats['sum_up'] + stats['sum_down']\n",
    "#     stats['CC2344_total_dif'] = stats['difference'].iloc[[13, 27]].sum()\n",
    "#     stats['CC2344_total_avg'] = stats['difference'].iloc[[13, 27]].mean()\n",
    "# stats['ALL_total_dif'] = stats['difference'].sum()\n",
    "# stats['ALL_total_avg'] = stats['difference'].mean()\n",
    "# stats.head(n=32)\n",
    "\n",
    "## LOG2FOLD STATS\n",
    "    \n",
    "# for i in CC2931_res:\n",
    "#     stats.at[i, 'mean_log2_all'] = (CC2931_log2Fold[i].abs()).mean(axis = 0, skipna = True)\n",
    "#     temp = CC2931_log2Fold[i].mask(CC2931_log2Fold[i] < 0)\n",
    "#     stats.at[i, 'mean_log2_up'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_up'] = temp.count()\n",
    "#     stats.at[i, 'percent_up'] = (temp.count()/CC2931_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2931_log2Fold[i].mask(CC2931_log2Fold[i] < 1.5)\n",
    "#     stats.at[i, 'mean_log2_above_1.5'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_above_1.5'] = temp.count()\n",
    "#     stats.at[i, 'percent_above_1.5'] = (temp.count()/CC2931_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2931_log2Fold[i].mask(CC2931_log2Fold[i] < 2)\n",
    "#     stats.at[i, 'mean_log2_above_2'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_above_2'] = temp.count()\n",
    "#     stats.at[i, 'percent_above_2'] = (temp.count()/CC2931_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2931_log2Fold[i].mask(CC2931_log2Fold[i] > 0)\n",
    "#     stats.at[i, 'mean_log2_down'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_down'] = temp.count()\n",
    "#     stats.at[i, 'percent_down'] = (temp.count()/CC2931_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2931_log2Fold[i].mask(CC2931_log2Fold[i] > -1.5)\n",
    "#     stats.at[i, 'mean_log2_below_-1.5'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_below_-1.5'] = temp.count()\n",
    "#     stats.at[i, 'percent_below_-1.5'] = (temp.count()/CC2931_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2931_log2Fold[i].mask(CC2931_log2Fold[i] > -2)\n",
    "#     stats.at[i, 'mean_log2_below_-2'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_below_-2'] = temp.count()\n",
    "#     stats.at[i, 'percent_below_-2'] = (temp.count()/CC2931_log2Fold[i].count())*100\n",
    "\n",
    "# for i in CC2344_res:\n",
    "#     stats.at[i, 'mean_log2_all'] = (CC2344_log2Fold[i].abs()).mean(axis = 0, skipna = True)\n",
    "#     temp = CC2344_log2Fold[i].mask(CC2344_log2Fold[i] < 0)\n",
    "#     stats.at[i, 'mean_log2_up'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_up'] = temp.count()\n",
    "#     stats.at[i, 'percent_up'] = (temp.count()/CC2344_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2344_log2Fold[i].mask(CC2344_log2Fold[i] < 1.5)\n",
    "#     stats.at[i, 'mean_log2_above_1.5'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_above_1.5'] = temp.count()\n",
    "#     stats.at[i, 'percent_above_1.5'] = (temp.count()/CC2344_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2344_log2Fold[i].mask(CC2344_log2Fold[i] < 2)\n",
    "#     stats.at[i, 'mean_log2_above_2'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_above_2'] = temp.count()\n",
    "#     stats.at[i, 'percent_above_2'] = (temp.count()/CC2344_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2344_log2Fold[i].mask(CC2344_log2Fold[i] > 0)\n",
    "#     stats.at[i, 'mean_log2_down'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_down'] = temp.count()\n",
    "#     stats.at[i, 'percent_down'] = (temp.count()/CC2344_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2344_log2Fold[i].mask(CC2344_log2Fold[i] > -1.5)\n",
    "#     stats.at[i, 'mean_log2_below_-1.5'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_below_-1.5'] = temp.count()\n",
    "#     stats.at[i, 'percent_below_-1.5'] = (temp.count()/CC2344_log2Fold[i].count())*100\n",
    "    \n",
    "#     temp = CC2344_log2Fold[i].mask(CC2344_log2Fold[i] > -2)\n",
    "#     stats.at[i, 'mean_log2_below_-2'] = temp.mean(axis = 0, skipna = True)\n",
    "#     stats.at[i, 'count_below_-2'] = temp.count()\n",
    "#     stats.at[i, 'percent_below_-2'] = (temp.count()/CC2344_log2Fold[i].count())*100"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
