{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "common-tuition",
   "metadata": {},
   "source": [
    "### KEYS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "macro-ranking",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy import random\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "from matplotlib.figure import Figure\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "from scipy import stats\n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy.stats import fligner\n",
    "from scipy.stats import skewtest\n",
    "stats.junk = lambda chisq, df: stats.chi2.sf(chisq, df)\n",
    "import csv\n",
    "import gffpandas.gffpandas as gffpd\n",
    "import sklearn\n",
    "from sklearn.metrics import r2_score\n",
    "import os\n",
    "import statistics\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "#### DATAFRAME RECORDING THE GENERATION TIME PER SAMPLE ####\n",
    "dic_gen = {'CC2344-L1': 912.356113, 'CC2344-L10': 917.129696, 'CC2344-L11': 889.5859554, 'CC2344-L12': 950.0552184, 'CC2344-L13': 961.4186064,\n",
    "           'CC2344-L14': 931.447801, 'CC2344-L15': 946.6643063, 'CC2344-L2': 923.1078072, 'CC2344-L3': 1000.469526, 'CC2344-L4': 808.9505794,\n",
    "           'CC2344-L5': 957.6380465, 'CC2344-L6': 970.6307256, 'CC2344-L7': 990.9451516, 'CC2344-L8': 1009.966123, 'CC2344-L9': 901.0619061, \n",
    "           'CC2931-L1': 1050.109001, 'CC2931-L10': 1097.978141, 'CC2931-L11': 1021.13559, 'CC2931-L13': 1041.362593, \n",
    "           'CC2931-L14': 1016.111493, 'CC2931-L15': 1052.540951, 'CC2931-L2': 1056.765369, 'CC2931-L3': 1000.399127, 'CC2931-L4': 1011.411706,\n",
    "           'CC2931-L5': 993.8603657, 'CC2931-L6': 1083.095655, 'CC2931-L7': 1067.34507, 'CC2931-L9': 1079.236285}\n",
    "generations = pd.Series(dic_gen)\n",
    "\n",
    "#### DATAFRAME RECORDING THE NUMBER OF MUTATIONS PER SAMPLE ####\n",
    "dic_mut = {'CC2344-L1': 396, 'CC2344-L10': 59, 'CC2344-L11': 46, 'CC2344-L12': 74, 'CC2344-L13': 49, 'CC2344-L14': 46, 'CC2344-L15': 53, \n",
    "           'CC2344-L2': 80, 'CC2344-L3': 63, 'CC2344-L4': 24, 'CC2344-L5': 68, 'CC2344-L6': 38, 'CC2344-L7': 45, 'CC2344-L8': 75, 'CC2344-L9': 27, \n",
    "           'CC2931-L1': 89, 'CC2931-L10': 87, 'CC2931-L11': 85, 'CC2931-L13': 97, 'CC2931-L14': 79, 'CC2931-L15': 141, 'CC2931-L2': 123,\n",
    "           'CC2931-L3': 52, 'CC2931-L4': 100, 'CC2931-L5': 335, 'CC2931-L6': 84, 'CC2931-L7': 72, 'CC2931-L9': 113}\n",
    "mutations = pd.Series(dic_mut)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "successful-myanmar",
   "metadata": {},
   "source": [
    "### COMBINING UNNORMALIZED RAW HTSEQ COUNTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "periodic-westminster",
   "metadata": {},
   "outputs": [],
   "source": [
    "## NAMES OF GENE COUNT FILES ##\n",
    "keyword = 'genes_count'\n",
    "label = []\n",
    "for fname in os.listdir('/research/projects/chlamydomonas/MAexpression/data/gene_count/'):\n",
    "    if keyword in fname:\n",
    "        fname = fname.replace('_genes_count','')\n",
    "        if 'CC2344' in fname:\n",
    "            label.append(fname)\n",
    "            \n",
    "# CC2344 - UNNORMALIZED GENE COUNTS ##\n",
    "gene_count = pd.DataFrame()\n",
    "for i in label:\n",
    "    section = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/gene_count/' + i + '_genes_count', delimiter = '\\t', names = ['name', i])\n",
    "    gene_count = pd.concat([gene_count, section], axis = 1)\n",
    "gene_count = gene_count.loc[:,~gene_count.columns.duplicated()]\n",
    "gene_count = gene_count.drop(gene_count.index.values[-5:])\n",
    "gene_count = gene_count.set_index('name')\n",
    "\n",
    "## EXPORTING \n",
    "gene_count.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2344_raw.txt', sep = '\\t', index = True, header = True)\n",
    "\n",
    "##################################################################################################################################\n",
    "\n",
    "## NAMES OF GENE COUNT FILES ##\n",
    "keyword = 'genes_count'\n",
    "label = []\n",
    "for fname in os.listdir('/research/projects/chlamydomonas/MAexpression/data/gene_count/'):\n",
    "    if keyword in fname:\n",
    "        fname = fname.replace('_genes_count','')\n",
    "        if 'CC2931' in fname:\n",
    "            label.append(fname)\n",
    "\n",
    "## CC2931 - UNNORMALIZED GENE COUNTS ##\n",
    "gene_count = pd.DataFrame()\n",
    "for i in label:\n",
    "    section = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/gene_count/' + i + '_genes_count', delimiter = '\\t', names = ['name', i])\n",
    "    gene_count = pd.concat([gene_count, section], axis = 1)\n",
    "    \n",
    "gene_count = gene_count.loc[:,~gene_count.columns.duplicated()]\n",
    "gene_count = gene_count.drop(gene_count.index.values[-5:])\n",
    "gene_count = gene_count.set_index('name')\n",
    "\n",
    "## EXPORTING\n",
    "gene_count.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2931_raw.txt', sep = '\\t', index = True, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "offensive-morocco",
   "metadata": {},
   "source": [
    "### GATHERING FRAGMENT LENGTHS FROM GFF FILE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "charged-particle",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/v6_genome_plus_anno/CC4532.v1_1.gene_exons.gff3', skiprows=3, delimiter = '\\t')\n",
    "gff.columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "gff[['attributes1', 'attributes2', 'attributes3', 'attributes4']] = gff['attributes'].str.split(';', n = 3, expand = True)\n",
    "gff_exon = gff.loc[gff['type'].isin(['mRNA', 'exon'])]\n",
    "gff_exon[['attributes1', 'attributes2', 'attributes4']] = gff_exon[['attributes1', 'attributes2', 'attributes4']].replace({'Parent=':'', 'ID=':''}, regex = True)\n",
    "gff_exon_pos = gff_exon.loc[gff_exon['strand'] == '+']\n",
    "gff_exon_neg = gff_exon.loc[gff_exon['strand'] == '-']\n",
    "mRNA = gff_exon.loc[gff_exon['type'] == 'mRNA'].reset_index()\n",
    "mRNA = mRNA.drop('index', axis = 1)\n",
    "mRNA['attributes'] = mRNA['attributes'].str.split(';')\n",
    "mRNA['parent'] = 'nan'\n",
    "for i in list(mRNA.index.values):\n",
    "    for a in mRNA.at[i,'attributes']:\n",
    "        if 'Parent=' in a:\n",
    "            mRNA.at[i, 'parent'] = a\n",
    "\n",
    "## EXTRACTING THE TOTAL SUM OF EXONS PER GENE\n",
    "mRNA['total_exon_length'] = 'nan'\n",
    "mRNA.rename(columns={'attributes1': 'ID'}, inplace = True)\n",
    "mRNA_ID = mRNA['ID'].values.tolist()\n",
    "gff_exon = gff_exon.loc[gff_exon['type'] == 'exon']\n",
    "\n",
    "for i in range(len(mRNA_ID)):\n",
    "    gene_length = []\n",
    "    section = gff_exon.loc[gff_exon['attributes2'] == mRNA_ID[i]]\n",
    "    for a in list(section.index.values):\n",
    "        gene_length.append(section.at[a,'end'] - section.at[a, 'start'])\n",
    "    mRNA.at[i, 'total_exon_length'] = sum(gene_length)\n",
    "mRNA_1 = mRNA.drop_duplicates(subset=['parent'], keep='last')\n",
    "\n",
    "fragment_length = pd.Series(data = mRNA_1['total_exon_length'].values.tolist(), index = list(parent), name = 'total_exon_length')\n",
    "\n",
    "## EXPORT DATA\n",
    "fragment_length.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/fpkm/fragment_length.csv', sep = '\\t', index = True, index_label= 'index', header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "metallic-signal",
   "metadata": {},
   "source": [
    "### INSTALLING PACKAGES AND LOADING LIBRARIES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "confidential-serum",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "## TIDYVERSE\n",
    "library(tidyverse)\n",
    "\n",
    "## DESEQ2\n",
    "library(\"BiocParallel\")\n",
    "register(MulticoreParam(4))\n",
    "library(\"DESeq2\")\n",
    "\n",
    "## CQN\n",
    "library(cqn)\n",
    "library(scales)\n",
    "\n",
    "## NOISEQ\n",
    "library(NOISeq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "smoking-garage",
   "metadata": {},
   "source": [
    "## NOISEQ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "initial-major",
   "metadata": {},
   "source": [
    "### PRE-FILTERING THROUGH NOISEQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "final-kennedy",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "#######################################\n",
    "### NOISEQ PRE-FILTERING FOR CC2931 ###\n",
    "#######################################\n",
    "CC2931 <- data.frame(read.csv('/research//projects//chlamydomonas//MAexpression//analysis//raw_counts//unnormalized_CC2931_raw.txt', sep = '\\t', header = TRUE, row.names = 'name', stringsAsFactors = FALSE))\n",
    "gene_length <- data.frame(read.csv('/research/projects/chlamydomonas/MAexpression/analysis/fpkm/fragment_length.csv', sep = '\\t', header = TRUE, stringsAsFactors = FALSE))\n",
    "factor <- colnames(CC2931)[1:length(CC2931)]\n",
    "factor[grep('ANC', factor)] <- 'control'\n",
    "factor[grep('L', factor)] <- 'treatment'\n",
    "factor <- matrix(factor, nrow = length(CC2931), ncol = 1)\n",
    "filtered_CC2931 <- filtered.data(CC2931, factor, norm=FALSE, method=2, p.adj='BH')\n",
    "\n",
    "#######################################\n",
    "### NOISEQ PRE-FILTERING FOR CC2344 ###\n",
    "#######################################\n",
    "CC2344 <- data.frame(read.table('/research//projects//chlamydomonas//MAexpression//analysis//raw_counts//unnormalized_CC2344_raw.txt', sep = '\\t', header = TRUE, row.names = 'name', stringsAsFactors = FALSE))\n",
    "factor <- colnames(CC2344)[1:length(CC2344)]\n",
    "factor[grep('ANC', factor)] <- 'control'\n",
    "factor[grep('L', factor)] <- 'treatment'\n",
    "factor <- matrix(factor, nrow = length(CC2344), ncol = 1)\n",
    "filtered_CC2344 <- filtered.data(CC2344, factor, norm=FALSE, method=2, p.adj='BH')\n",
    "\n",
    "## EXPORTING PREFILTERED TEXTFILE\n",
    "write.table(as.data.frame(filtered_CC2931), 'raw_counts//noiseq_filtered_unnormalized_CC2931.txt', sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(filtered_CC2344), 'raw_counts//noiseq_filtered_unnormalized_CC2344.txt', sep=\"\\t\", quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "realistic-contribution",
   "metadata": {},
   "source": [
    "### FILTERED GC CONTENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "leading-storm",
   "metadata": {},
   "outputs": [],
   "source": [
    "## OPENING GC LENGTH AND FRAGMENT LENGTH\n",
    "gc_content = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/genome_info/v6_genome_plus_anno/GC_lengths.tsv', delimiter = '\\t')\n",
    "fragment_length = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/fpkm/fragment_length.csv', delimiter = '\\t')\n",
    "fragment_length = fragment_length.set_index('Unnamed: 0')\n",
    "\n",
    "# OPENING NOISEQ FILTERED GENE COUNTS\n",
    "CC2931_gene_count = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/noiseq_filtered_unnormalized_CC2931.txt', delimiter = '\\t')\n",
    "CC2344_gene_count = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/noiseq_filtered_unnormalized_CC2344.txt', delimiter = '\\t')\n",
    "\n",
    "CC2931_gene_count = CC2931_gene_count.join([fragment_length, gc_content['GC']])\n",
    "CC2344_gene_count = CC2344_gene_count.join([fragment_length, gc_content['GC']])\n",
    "\n",
    "# EXPORT ##\n",
    "CC2344_gene_count.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2344_filtered_noiseq_wt_GC', sep = '\\t', index = True, header = True)\n",
    "CC2931_gene_count.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2931_filtered_noiseq_wt_GC', sep = '\\t', index = True, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "informational-sunrise",
   "metadata": {},
   "source": [
    "### DESEQ NORMALIZING NOISEQ FILTERED DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "norwegian-insured",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "################\n",
    "#### CC2931 ####\n",
    "################\n",
    "CC2931 <- data.frame(read.csv('/research//projects//chlamydomonas//MAexpression//analysis//raw_counts//CC2931_filtered_noiseq_wt_GC', sep = '\\t', header = TRUE, row.names = 'X', stringsAsFactors = FALSE))\n",
    "condition <- colnames(CC2931)[1:(length(CC2931)-2)]\n",
    "counts <- CC2931[condition]\n",
    "condition_1 <- condition\n",
    "condition[grep('ANC', condition)] <- 'control'\n",
    "condition[grep('L', condition)] <- 'treatment'\n",
    "\n",
    "coldata <- data.frame(condition, condition_1, row.names = 'condition_1')\n",
    "\n",
    "dds <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ condition)\n",
    "directory <- \"/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/\"\n",
    "dds <- estimateSizeFactors(dds)\n",
    "counts <- counts(dds, normalized=TRUE)\n",
    "write.table(counts, paste(directory, \"CC2931_noiseq_filtered_deseq_normalized\", \".txt\", sep=\"\"), sep=\"\\t\", quote = F)\n",
    "\n",
    "################\n",
    "#### CC2344 ####\n",
    "################\n",
    "CC2344 <- data.frame(read.csv('/research//projects//chlamydomonas//MAexpression//analysis//raw_counts//CC2344_filtered_noiseq_wt_GC', sep = '\\t', header = TRUE, row.names = 'X', stringsAsFactors = FALSE))\n",
    "condition <- colnames(CC2344)[1:(length(CC2344)-2)]\n",
    "counts <- CC2344[condition]\n",
    "condition_1 <- condition\n",
    "condition[grep('ANC', condition)] <- 'control'\n",
    "condition[grep('L', condition)] <- 'treatment'\n",
    "\n",
    "coldata <- data.frame(condition, condition_1, row.names = 'condition_1')\n",
    "\n",
    "dds <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ condition)\n",
    "directory <- \"/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/\"\n",
    "dds <- estimateSizeFactors(dds)\n",
    "counts <- counts(dds, normalized=TRUE)\n",
    "write.table(counts, paste(directory, \"CC2344_noiseq_filtered_deseq_normalized\", \".txt\", sep=\"\"), sep=\"\\t\", quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "affecting-difficulty",
   "metadata": {},
   "source": [
    "### ADDING GC CONTENT TO DESEQ2 NORMALIZED NOISEQ FILTERED GENE COUNTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adapted-cuisine",
   "metadata": {},
   "outputs": [],
   "source": [
    "## OPENING DESEQ2 NORMALIZED NOISEQ FILTERED GENE COUNTS\n",
    "CC2931_normalized_GC = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2931_noiseq_filtered_deseq_normalized.txt', delimiter = '\\t')\n",
    "CC2344_normalized_GC = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2344_noiseq_filtered_deseq_normalized.txt', delimiter = '\\t')\n",
    "\n",
    "CC2931_normalized_GC = CC2931_normalized_GC.join([fragment_length, gc_content['GC']])\n",
    "CC2344_normalized_GC = CC2344_normalized_GC.join([fragment_length, gc_content['GC']])\n",
    "\n",
    "## EXPORT ##\n",
    "CC2931_normalized_GC.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2931_deseq_normalized_noiseq_filtered_wt_GC', sep = '\\t', index = True, header = True)\n",
    "CC2344_normalized_GC.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2344_deseq_normalized_noiseq_filtered_wt_GC', sep = '\\t', index = True, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wound-uganda",
   "metadata": {},
   "source": [
    "### CREATING NOISEQ DATA OBJECT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "streaming-rebecca",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#####################################\n",
    "### CREATING A NOISEQ DATA OBJECT ###\n",
    "#####################################\n",
    "CC2931_deseq_normalized <- data.frame(read.table('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2931_deseq_normalized_noiseq_filtered_wt_GC', sep = '\\t', header = TRUE, stringsAsFactors = FALSE))\n",
    "CC2931_length <- data.frame(CC2931_deseq_normalized$total_exon_length)\n",
    "CC2931_GC <- data.frame(CC2931_deseq_normalized$GC)\n",
    "\n",
    "## PROVIDING INDEXES\n",
    "row.names(CC2931_length) <- CC2931_deseq_normalized$X\n",
    "row.names(CC2931_GC) <- CC2931_deseq_normalized$X\n",
    "row.names(CC2931_deseq_normalized) <- CC2931_deseq_normalized$X\n",
    "\n",
    "## CREATING FACTORS AND COUNTS\n",
    "CC2931_counts <- CC2931_deseq_normalized[colnames(CC2931_deseq_normalized)[2:(length(CC2931_deseq_normalized)-2)]]\n",
    "CC2931_factor <- colnames(CC2931_counts)\n",
    "CC2931_columns <- CC2931_factor\n",
    "colnames(CC2931_counts) <- CC2931_columns\n",
    "CC2931_factor[grep('ANC', CC2931_factor)] <- 'control'\n",
    "CC2931_factor[grep('L', CC2931_factor)] <- 'treatment'\n",
    "num <- c(1:length(t(CC2931_factor)))\n",
    "CC2931_factor <- data.frame(num, CC2931_factor, CC2931_columns, row.names = 'num')\n",
    "##############################################################################################################################\n",
    "\n",
    "CC2344_deseq_normalized <- data.frame(read.table('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2344_deseq_normalized_noiseq_filtered_wt_GC', sep = '\\t', header = TRUE, stringsAsFactors = FALSE))\n",
    "CC2344_length <- data.frame(CC2344_deseq_normalized$total_exon_length)\n",
    "CC2344_GC <- data.frame(CC2344_deseq_normalized$GC)\n",
    "\n",
    "## PROVIDING INDEXES\n",
    "row.names(CC2344_length) <- CC2344_deseq_normalized$X\n",
    "row.names(CC2344_GC) <- CC2344_deseq_normalized$X\n",
    "row.names(CC2344_deseq_normalized) <- CC2344_deseq_normalized$X\n",
    "\n",
    "## CREATING FACTORS AND COUNTS\n",
    "CC2344_counts <- CC2344_deseq_normalized[colnames(CC2344_deseq_normalized)[2:(length(CC2344_deseq_normalized)-2)]]\n",
    "CC2344_factor <- colnames(CC2344_counts)\n",
    "CC2344_columns <- CC2344_factor\n",
    "colnames(CC2344_counts) <- CC2344_columns\n",
    "CC2344_factor[grep('ANC', CC2344_factor)] <- 'control'\n",
    "CC2344_factor[grep('L', CC2344_factor)] <- 'treatment'\n",
    "CC2344_factor <- data.frame(CC2344_factor)\n",
    "\n",
    "\n",
    "## CREATING READDATA FILES\n",
    "data_CC2931 <- readData(data = CC2931_counts, length = CC2931_length, gc = CC2931_GC, CC2931_factor)\n",
    "data_CC2344 <- readData(data = CC2344_counts, length = CC2344_length, gc = CC2344_GC, factors = CC2344_factor)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "enclosed-bleeding",
   "metadata": {},
   "source": [
    "## CONDITIONAL QUANTILE NORMALIZATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "regulated-transport",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################\n",
    "### CC2931 - CQN NORMALIZATION ####\n",
    "###################################\n",
    "directory <- \"/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/\"\n",
    "CC2931_noiseq_filtered <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2931_filtered_noiseq_wt_GC', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = 'X')\n",
    "counts <- CC2931_noiseq_filtered[c(1:42)]\n",
    "cqn_res <- cqn(counts = counts, x = CC2931_noiseq_filtered$GC, lengths = CC2931_noiseq_filtered$total_exon_length)\n",
    "\n",
    "#### CREATING FACTORS AND COLDATA ####\n",
    "condition <- c('ANC', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7', 'L9', 'L10', 'L11', 'L13', 'L14', 'L15')\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "for (i in condition) {factor[grep((i), factor)] <- i}\n",
    "factor[grep('L1.rep', factor)] <- 'L1'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### CREATING PLOT ####\n",
    "jpeg(file=\"cqn/CC2931_gc_plot_before.jpeg\")\n",
    "cqnplot(cqn_res, n = 1, xlab = 'GC content', ylim = c(5,10))\n",
    "dev.off()\n",
    "\n",
    "jpeg(file=\"cqn/CC2931_length_plot_before.jpeg\")\n",
    "cqnplot(cqn_res, n = 2, xlim = c(-2,4.5), xlab = 'length', ylim = c(5.5,8.5))\n",
    "dev.off()\n",
    "\n",
    "#### NORMALIZED EXPRESSION VALUES ####\n",
    "cqnnorm <- cqn_res$y + cqn_res$offset\n",
    "cqnOffset <- cqn_res$glm.offset\n",
    "cqnNormFactors <- exp(cqnOffset)\n",
    "cqnNormFactors[is.na(cqnNormFactors)] <- 1 ## replacing missing values with 1\n",
    "normFactors_sameScale <- cqnNormFactors/exp(rowMeans(log(cqnNormFactors)))\n",
    "\n",
    "#### CREATE DATASET USING NORMFACTORS_SAMESCALE ####\n",
    "CC2931_dds1 <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)\n",
    "normalizationFactors(CC2931_dds1) <- normFactors_sameScale\n",
    "counts_nobias_sameScale = counts(CC2931_dds1, normalized = TRUE)\n",
    "write.table(counts_nobias_sameScale, paste(directory, \"CC2931_noiseq_filtered_cqn_normalized_samescale\", \".txt\", sep=\"\"), sep=\"\\t\", quote = F)\n",
    "\n",
    "#### CREATE DATASET USING NORMFACTORS ####\n",
    "CC2931_dds2 <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)\n",
    "normalizationFactors(CC2931_dds2) = cqnNormFactors\n",
    "counts_nobias = counts(CC2931_dds2, normalized = TRUE)\n",
    "write.table(counts_nobias, paste(directory, \"CC2931_noiseq_filtered_cqn_normalized\", \".txt\", sep=\"\"), sep=\"\\t\", quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "about-holly",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################\n",
    "#### CC2344 - CQN NORMALIZATION ####\n",
    "####################################\n",
    "directory <- \"/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/\"\n",
    "CC2344_noiseq_filtered <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2344_filtered_noiseq_wt_GC', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = 'X')\n",
    "counts <- CC2344_noiseq_filtered[c(1:48)]\n",
    "cqn_res <- cqn(counts = counts, x = CC2344_noiseq_filtered$GC, lengths = CC2344_noiseq_filtered$total_exon_length)\n",
    "\n",
    "#### CREATING FACTORS AND COLDATA ####\n",
    "condition <- c('ANC', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7', 'L8', 'L9', 'L10', 'L11', 'L12', 'L13', 'L14', 'L15')\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "for (i in condition) {factor[grep((i), factor)] <- i}\n",
    "factor[grep('L1.rep', factor)] <- 'L1'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### CREATING PLOT ####\n",
    "jpeg(file = \"cqn/CC2344_gc_plot_before.jpeg\")\n",
    "cqnplot(cqn_res, n = 1, xlab = 'GC content', ylim = c(3, 10))\n",
    "dev.off()\n",
    "\n",
    "jpeg(file = \"cqn/CC2344_length_plot_before.jpeg\")\n",
    "cqnplot(cqn_res, n = 2, xlim = c(-2,4.5), xlab = 'length', ylim = c(4, 8))\n",
    "dev.off()\n",
    "\n",
    "#### NORMALIZED EXPRESSION VALUES ####\n",
    "cqnnorm <- cqn_res$y + cqn_res$offset\n",
    "cqnOffset <- cqn_res$glm.offset\n",
    "cqnNormFactors <- exp(cqnOffset)\n",
    "cqnNormFactors[is.na(cqnNormFactors)] <- 1 ## replacing missing values with 1\n",
    "normFactors_sameScale <- cqnNormFactors/exp(rowMeans(log(cqnNormFactors)))\n",
    "\n",
    "#### CREATE DATASET USING NORMFACTORS_SAMESCALE ####\n",
    "CC2344_dds1 <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)\n",
    "normalizationFactors(CC2344_dds1) <- normFactors_sameScale\n",
    "counts_nobias_sameScale = counts(CC2344_dds1, normalized = TRUE)\n",
    "write.table(counts_nobias_sameScale, paste(directory, \"CC2344_noiseq_filtered_cqn_normalized_samescale\", \".txt\", sep=\"\"), sep=\"\\t\", quote = F)\n",
    "\n",
    "#### CREATE DATASET USING NORMFACTORS ####\n",
    "CC2344_dds2 <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)\n",
    "normalizationFactors(CC2344_dds2) = cqnNormFactors\n",
    "counts_nobias = counts(CC2344_dds2, normalized = TRUE)\n",
    "write.table(counts_nobias, paste(directory, \"CC2344_noiseq_filtered_cqn_normalized\", \".txt\", sep=\"\"), sep=\"\\t\", quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "opposite-patrick",
   "metadata": {},
   "outputs": [],
   "source": [
    "################################################\n",
    "### CC2931 - RECHECKING CQN NORMALIZED DATA ####\n",
    "################################################\n",
    "CC2931_cqn_filtered <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2931_noiseq_filtered_cqn_normalized_samescale.txt', sep = '\\t', header = TRUE, stringsAsFactors = FALSE)\n",
    "CC2931_noiseq_filtered <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2931_filtered_noiseq_wt_GC', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = 'X')\n",
    "cqn_res <- cqn(counts = CC2931_cqn_filtered, x = CC2931_noiseq_filtered$GC, lengths = CC2931_noiseq_filtered$total_exon_length)\n",
    "\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "factor[grep('ANC', factor)] <- 'control'\n",
    "factor[grep('L', factor)] <- 'treatment'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### CREATING PLOT ####\n",
    "jpeg(file = \"cqn/CC2931_gc_plot_after.jpeg\")\n",
    "cqnplot(cqn_res, n = 1, xlab = 'GC content', ylim = c(5,10))\n",
    "dev.off()\n",
    "\n",
    "jpeg(file = \"cqn/CC2931_length_plot_after.jpeg\")\n",
    "cqnplot(cqn_res, n = 2, xlim = c(-2,4.5), xlab = 'length', ylim = c(5.5, 8.5))\n",
    "dev.off()\n",
    "\n",
    "################################################\n",
    "### CC2344 - RECHECKING CQN NORMALIZED DATA ####\n",
    "################################################\n",
    "CC2344_cqn_filtered <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2344_noiseq_filtered_cqn_normalized_samescale.txt', sep = '\\t', header = TRUE, stringsAsFactors = FALSE)\n",
    "CC2344_noiseq_filtered <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2344_filtered_noiseq_wt_GC', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = 'X')\n",
    "cqn_res <- cqn(counts = CC2344_cqn_filtered, x = CC2344_noiseq_filtered$GC, lengths = CC2344_noiseq_filtered$total_exon_length)\n",
    "\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "factor[grep('ANC', factor)] <- 'control'\n",
    "factor[grep('L', factor)] <- 'treatment'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### CREATING PLOT ####\n",
    "jpeg(file = \"cqn/CC2344_gc_plot_after.jpeg\")\n",
    "cqnplot(cqn_res, n = 1, xlab = 'GC content', ylim = c(3,10))\n",
    "dev.off()\n",
    "\n",
    "jpeg(file = \"cqn/CC2344_length_plot_after.jpeg\")\n",
    "cqnplot(cqn_res, n = 2, xlim = c(-2,4.5), xlab = 'length', ylim = c(4,8))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "healthy-holly",
   "metadata": {},
   "source": [
    "## DESEQ2 - DIFFERENTIAL EXPRESSION ANALYSIS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "third-protocol",
   "metadata": {},
   "source": [
    "### CC2931"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "verified-police",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### USING DESEQ #####\n",
    "dds <- DESeq(CC2931_dds1, parallel = TRUE)\n",
    "res_L1 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L1'), alpha = 0.05)\n",
    "res_L2 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L2'), alpha = 0.05)\n",
    "res_L3 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L3'), alpha = 0.05)\n",
    "res_L4 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L4'), alpha = 0.05)\n",
    "res_L5 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L5'), alpha = 0.05)\n",
    "res_L6 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L6'), alpha = 0.05)\n",
    "res_L7 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L7'), alpha = 0.05)\n",
    "res_L9 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L9'), alpha = 0.05)\n",
    "res_L10 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L10'), alpha = 0.05)\n",
    "res_L11 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L11'), alpha = 0.05)\n",
    "res_L13 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L13'), alpha = 0.05)\n",
    "res_L14 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L14'), alpha = 0.05)\n",
    "res_L15 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L15'), alpha = 0.05)\n",
    "\n",
    "#### SUMMARY OF PROFILE ####\n",
    "resOrdered_L1 <- res_L1[order(res_L1$pvalue),]\n",
    "resOrdered_L2 <- res_L2[order(res_L2$pvalue),]\n",
    "resOrdered_L3 <- res_L3[order(res_L3$pvalue),]\n",
    "resOrdered_L4 <- res_L4[order(res_L4$pvalue),]\n",
    "resOrdered_L5 <- res_L5[order(res_L5$pvalue),]\n",
    "resOrdered_L6 <- res_L6[order(res_L6$pvalue),]\n",
    "resOrdered_L7 <- res_L7[order(res_L7$pvalue),]\n",
    "resOrdered_L9 <- res_L9[order(res_L9$pvalue),]\n",
    "resOrdered_L10 <- res_L10[order(res_L10$pvalue),]\n",
    "resOrdered_L11 <- res_L11[order(res_L11$pvalue),]\n",
    "resOrdered_L13 <- res_L13[order(res_L13$pvalue),]\n",
    "resOrdered_L14 <- res_L14[order(res_L14$pvalue),]\n",
    "resOrdered_L15 <- res_L15[order(res_L15$pvalue),]\n",
    "\n",
    "# summary(resOrdered_L1)\n",
    "# summary(resOrdered_L2)\n",
    "# summary(resOrdered_L3)\n",
    "# summary(resOrdered_L4)\n",
    "# summary(resOrdered_L5)\n",
    "# summary(resOrdered_L6)\n",
    "# summary(resOrdered_L7)\n",
    "# summary(resOrdered_L9)\n",
    "# summary(resOrdered_L10)\n",
    "# summary(resOrdered_L11)\n",
    "# summary(resOrdered_L13)\n",
    "# summary(resOrdered_L14)\n",
    "# summary(resOrdered_L15)\n",
    "\n",
    "# #### LFC SHRINKAGE ####\n",
    "# resLFC_L1 <- lfcShrink(dds, coef=\"factor_L1_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L2 <- lfcShrink(dds, coef=\"factor_L2_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L3 <- lfcShrink(dds, coef=\"factor_L3_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L4 <- lfcShrink(dds, coef=\"factor_L4_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L5 <- lfcShrink(dds, coef=\"factor_L5_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L6 <- lfcShrink(dds, coef=\"factor_L6_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L7 <- lfcShrink(dds, coef=\"factor_L7_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L9 <- lfcShrink(dds, coef=\"factor_L9_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L10 <- lfcShrink(dds, coef=\"factor_L10_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L11 <- lfcShrink(dds, coef=\"factor_L11_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L13 <- lfcShrink(dds, coef=\"factor_L13_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L14 <- lfcShrink(dds, coef=\"factor_L14_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L15 <- lfcShrink(dds, coef=\"factor_L15_vs_ANC\", type=\"apeglm\")\n",
    "\n",
    "#### FILES OF DIFFERNTIALLY EXPRESSED GENES ####\n",
    "resSig_L1 = resOrdered_L1[which(resOrdered_L1$padj < 0.05),]\n",
    "resSig_L2 = resOrdered_L2[which(resOrdered_L2$padj < 0.05),]\n",
    "resSig_L3 = resOrdered_L3[which(resOrdered_L3$padj < 0.05),]\n",
    "resSig_L4 = resOrdered_L4[which(resOrdered_L4$padj < 0.05),]\n",
    "resSig_L5 = resOrdered_L5[which(resOrdered_L5$padj < 0.05),]\n",
    "resSig_L6 = resOrdered_L6[which(resOrdered_L6$padj < 0.05),]\n",
    "resSig_L7 = resOrdered_L7[which(resOrdered_L7$padj < 0.05),]\n",
    "resSig_L9 = resOrdered_L9[which(resOrdered_L9$padj < 0.05),]\n",
    "resSig_L10 = resOrdered_L10[which(resOrdered_L10$padj < 0.05),]\n",
    "resSig_L11 = resOrdered_L11[which(resOrdered_L11$padj < 0.05),]\n",
    "resSig_L13 = resOrdered_L13[which(resOrdered_L13$padj < 0.05),]\n",
    "resSig_L14 = resOrdered_L14[which(resOrdered_L14$padj < 0.05),]\n",
    "resSig_L15 = resOrdered_L15[which(resOrdered_L15$padj < 0.05),]\n",
    "\n",
    "summary(resSig_L1)\n",
    "summary(resSig_L2)\n",
    "summary(resSig_L3)\n",
    "summary(resSig_L4)\n",
    "summary(resSig_L5)\n",
    "summary(resSig_L6)\n",
    "summary(resSig_L7)\n",
    "summary(resSig_L9)\n",
    "summary(resSig_L10)\n",
    "summary(resSig_L11)\n",
    "summary(resSig_L13)\n",
    "summary(resSig_L14)\n",
    "summary(resSig_L15)\n",
    "\n",
    "# #### PLOTS ####\n",
    "# plotMA(res_L1, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L1, ylim=c(-5,5))\n",
    "# plotMA(res_L2, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L2, ylim=c(-5,5))\n",
    "# plotMA(res_L3, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L3, ylim=c(-5,5))\n",
    "# plotMA(res_L4, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L4, ylim=c(-5,5))\n",
    "# plotMA(res_L5, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L5, ylim=c(-5,5))\n",
    "# plotMA(res_L6, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L6, ylim=c(-5,5))\n",
    "# plotMA(res_L7, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L7, ylim=c(-5,5))\n",
    "# plotMA(res_L9, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L9, ylim=c(-5,5))\n",
    "# plotMA(res_L10, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L10, ylim=c(-5,5))\n",
    "# plotMA(res_L11, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L11, ylim=c(-5,5))\n",
    "# plotMA(res_L13, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L13, ylim=c(-5,5))\n",
    "# plotMA(res_L14, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L14, ylim=c(-5,5))\n",
    "# plotMA(res_L15, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L15, ylim=c(-5,5))\n",
    "# abline(h=c(-1,1), col=\"dodgerblue\", lwd=2)\n",
    "\n",
    "#### EXPORTING FILES OF DIFFERNTIALLY EXPRESSED GENES ####\n",
    "write.csv(as.data.frame(resSig_L1), file=\"CC2931.resSig_L1.genes\")\n",
    "write.csv(as.data.frame(resSig_L2), file=\"CC2931.resSig_L2.genes\")\n",
    "write.csv(as.data.frame(resSig_L3), file=\"CC2931.resSig_L3.genes\")\n",
    "write.csv(as.data.frame(resSig_L4), file=\"CC2931.resSig_L4.genes\")\n",
    "write.csv(as.data.frame(resSig_L5), file=\"CC2931.resSig_L5.genes\")\n",
    "write.csv(as.data.frame(resSig_L6), file=\"CC2931.resSig_L6.genes\")\n",
    "write.csv(as.data.frame(resSig_L7), file=\"CC2931.resSig_L7.genes\")\n",
    "write.csv(as.data.frame(resSig_L9), file=\"CC2931.resSig_L9.genes\")\n",
    "write.csv(as.data.frame(resSig_L10), file=\"CC2931.resSig_L10.genes\")\n",
    "write.csv(as.data.frame(resSig_L11), file=\"CC2931.resSig_L11.genes\")\n",
    "write.csv(as.data.frame(resSig_L13), file=\"CC2931.resSig_L13.genes\")\n",
    "write.csv(as.data.frame(resSig_L14), file=\"CC2931.resSig_L14.genes\")\n",
    "write.csv(as.data.frame(resSig_L15), file=\"CC2931.resSig_L15.genes\")\n",
    "\n",
    "#### EXPORTING LOG2FOLD CHANGE RESULTS ####\n",
    "write.table(as.data.frame(res_L1), 'genes_log2fold/CC2931-L1',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L2), 'genes_log2fold/CC2931-L2',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L3), 'genes_log2fold/CC2931-L3',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L4), 'genes_log2fold/CC2931-L4',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L5), 'genes_log2fold/CC2931-L5',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L6), 'genes_log2fold/CC2931-L6',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L7), 'genes_log2fold/CC2931-L7',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L9), 'genes_log2fold/CC2931-L9',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L10), 'genes_log2fold/CC2931-L10',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L11), 'genes_log2fold/CC2931-L11',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L13), 'genes_log2fold/CC2931-L13',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L14), 'genes_log2fold/CC2931-L14',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L15), 'genes_log2fold/CC2931-L15',  sep=\"\\t\", quote = F)\n",
    "\n",
    "# #### PCA PLOT ####\n",
    "# vsdata <- vst(dds, blind=FALSE)\n",
    "# plotPCA(vsdata, intgroup=\"factor\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "disabled-england",
   "metadata": {},
   "source": [
    "### CC2344"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "advance-football",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### USING DESEQ #####\n",
    "dds <- DESeq(CC2344_dds1, parallel = TRUE)\n",
    "res_L1 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L1'), alpha = 0.05)\n",
    "res_L2 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L2'), alpha = 0.05)\n",
    "res_L3 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L3'), alpha = 0.05)\n",
    "res_L4 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L4'), alpha = 0.05)\n",
    "res_L5 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L5'), alpha = 0.05)\n",
    "res_L6 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L6'), alpha = 0.05)\n",
    "res_L7 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L7'), alpha = 0.05)\n",
    "res_L8 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L8'), alpha = 0.05)\n",
    "res_L9 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L9'), alpha = 0.05)\n",
    "res_L10 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L10'), alpha = 0.05)\n",
    "res_L11 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L11'), alpha = 0.05)\n",
    "res_L12 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L12'), alpha = 0.05)\n",
    "res_L13 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L13'), alpha = 0.05)\n",
    "res_L14 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L14'), alpha = 0.05)\n",
    "res_L15 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L15'), alpha = 0.05)\n",
    "\n",
    "#### SUMMARY OF PROFILE ####\n",
    "resOrdered_L1 <- res_L1[order(res_L1$pvalue),]\n",
    "resOrdered_L2 <- res_L2[order(res_L2$pvalue),]\n",
    "resOrdered_L3 <- res_L3[order(res_L3$pvalue),]\n",
    "resOrdered_L4 <- res_L4[order(res_L4$pvalue),]\n",
    "resOrdered_L5 <- res_L5[order(res_L5$pvalue),]\n",
    "resOrdered_L6 <- res_L6[order(res_L6$pvalue),]\n",
    "resOrdered_L7 <- res_L7[order(res_L7$pvalue),]\n",
    "resOrdered_L8 <- res_L8[order(res_L8$pvalue),]\n",
    "resOrdered_L9 <- res_L9[order(res_L9$pvalue),]\n",
    "resOrdered_L10 <- res_L10[order(res_L10$pvalue),]\n",
    "resOrdered_L11 <- res_L11[order(res_L11$pvalue),]\n",
    "resOrdered_L12 <- res_L12[order(res_L12$pvalue),]\n",
    "resOrdered_L13 <- res_L13[order(res_L13$pvalue),]\n",
    "resOrdered_L14 <- res_L14[order(res_L14$pvalue),]\n",
    "resOrdered_L15 <- res_L15[order(res_L15$pvalue),]\n",
    "\n",
    "# summary(resOrdered_L1)\n",
    "# summary(resOrdered_L2)\n",
    "# summary(resOrdered_L3)\n",
    "# summary(resOrdered_L4)\n",
    "# summary(resOrdered_L5)\n",
    "# summary(resOrdered_L6)\n",
    "# summary(resOrdered_L7)\n",
    "# summary(resOrdered_L8)\n",
    "# summary(resOrdered_L9)\n",
    "# summary(resOrdered_L10)\n",
    "# summary(resOrdered_L11)\n",
    "# summary(resOrdered_L12)\n",
    "# summary(resOrdered_L13)\n",
    "# summary(resOrdered_L14)\n",
    "# summary(resOrdered_L15)\n",
    "\n",
    "# #### LFC SHRINKAGE ####\n",
    "# resLFC_L1 <- lfcShrink(dds, coef=\"factor_L1_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L2 <- lfcShrink(dds, coef=\"factor_L2_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L3 <- lfcShrink(dds, coef=\"factor_L3_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L4 <- lfcShrink(dds, coef=\"factor_L4_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L5 <- lfcShrink(dds, coef=\"factor_L5_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L6 <- lfcShrink(dds, coef=\"factor_L6_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L7 <- lfcShrink(dds, coef=\"factor_L7_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L8 <- lfcShrink(dds, coef=\"factor_L8_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L9 <- lfcShrink(dds, coef=\"factor_L9_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L10 <- lfcShrink(dds, coef=\"factor_L10_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L11 <- lfcShrink(dds, coef=\"factor_L11_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L8 <- lfcShrink(dds, coef=\"factor_L12_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L13 <- lfcShrink(dds, coef=\"factor_L13_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L14 <- lfcShrink(dds, coef=\"factor_L14_vs_ANC\", type=\"apeglm\")\n",
    "# resLFC_L15 <- lfcShrink(dds, coef=\"factor_L15_vs_ANC\", type=\"apeglm\")\n",
    "\n",
    "#### FILES OF DIFFERNTIALLY EXPRESSED GENES ####\n",
    "resSig_L1 = resOrdered_L1[which(resOrdered_L1$padj < 0.05),]\n",
    "resSig_L2 = resOrdered_L2[which(resOrdered_L2$padj < 0.05),]\n",
    "resSig_L3 = resOrdered_L3[which(resOrdered_L3$padj < 0.05),]\n",
    "resSig_L4 = resOrdered_L4[which(resOrdered_L4$padj < 0.05),]\n",
    "resSig_L5 = resOrdered_L5[which(resOrdered_L5$padj < 0.05),]\n",
    "resSig_L6 = resOrdered_L6[which(resOrdered_L6$padj < 0.05),]\n",
    "resSig_L7 = resOrdered_L7[which(resOrdered_L7$padj < 0.05),]\n",
    "resSig_L8 = resOrdered_L8[which(resOrdered_L8$padj < 0.05),]\n",
    "resSig_L9 = resOrdered_L9[which(resOrdered_L9$padj < 0.05),]\n",
    "resSig_L10 = resOrdered_L10[which(resOrdered_L10$padj < 0.05),]\n",
    "resSig_L11 = resOrdered_L11[which(resOrdered_L11$padj < 0.05),]\n",
    "resSig_L12 = resOrdered_L12[which(resOrdered_L12$padj < 0.05),]\n",
    "resSig_L13 = resOrdered_L13[which(resOrdered_L13$padj < 0.05),]\n",
    "resSig_L14 = resOrdered_L14[which(resOrdered_L14$padj < 0.05),]\n",
    "resSig_L15 = resOrdered_L15[which(resOrdered_L15$padj < 0.05),]\n",
    "\n",
    "summary(resSig_L1)\n",
    "summary(resSig_L2)\n",
    "summary(resSig_L3)\n",
    "summary(resSig_L4)\n",
    "summary(resSig_L5)\n",
    "summary(resSig_L6)\n",
    "summary(resSig_L7)\n",
    "summary(resSig_L8)\n",
    "summary(resSig_L9)\n",
    "summary(resSig_L10)\n",
    "summary(resSig_L11)\n",
    "summary(resSig_L12)\n",
    "summary(resSig_L13)\n",
    "summary(resSig_L14)\n",
    "summary(resSig_L15)\n",
    "\n",
    "# #### PLOTS ####\n",
    "# plotMA(res_L1, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L1, ylim=c(-5,5))\n",
    "# plotMA(res_L2, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L2, ylim=c(-5,5))\n",
    "# plotMA(res_L3, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L3, ylim=c(-5,5))\n",
    "# plotMA(res_L4, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L4, ylim=c(-5,5))\n",
    "# plotMA(res_L5, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L5, ylim=c(-5,5))\n",
    "# plotMA(res_L6, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L6, ylim=c(-5,5))\n",
    "# plotMA(res_L7, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L7, ylim=c(-5,5))\n",
    "# plotMA(res_L8, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L8, ylim=c(-5,5))\n",
    "# plotMA(res_L9, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L9, ylim=c(-5,5))\n",
    "# plotMA(res_L10, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L10, ylim=c(-5,5))\n",
    "# plotMA(res_L11, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L11, ylim=c(-5,5))\n",
    "# plotMA(res_L12, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L12, ylim=c(-5,5))\n",
    "# plotMA(res_L13, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L13, ylim=c(-5,5))\n",
    "# plotMA(res_L14, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L14, ylim=c(-5,5))\n",
    "# plotMA(res_L15, ylim=c(-5,5))\n",
    "# plotMA(resLFC_L15, ylim=c(-5,5))\n",
    "# abline(h=c(-1,1), col=\"dodgerblue\", lwd=2)\n",
    "\n",
    "#### EXPORTING FILES OF DIFFERNTIALLY EXPRESSED GENES ####\n",
    "write.csv(as.data.frame(resSig_L1), file=\"CC2344.resSig_L1.genes\")\n",
    "write.csv(as.data.frame(resSig_L2), file=\"CC2344.resSig_L2.genes\")\n",
    "write.csv(as.data.frame(resSig_L3), file=\"CC2344.resSig_L3.genes\")\n",
    "write.csv(as.data.frame(resSig_L4), file=\"CC2344.resSig_L4.genes\")\n",
    "write.csv(as.data.frame(resSig_L5), file=\"CC2344.resSig_L5.genes\")\n",
    "write.csv(as.data.frame(resSig_L6), file=\"CC2344.resSig_L6.genes\")\n",
    "write.csv(as.data.frame(resSig_L7), file=\"CC2344.resSig_L7.genes\")\n",
    "write.csv(as.data.frame(resSig_L8), file=\"CC2344.resSig_L8.genes\")\n",
    "write.csv(as.data.frame(resSig_L9), file=\"CC2344.resSig_L9.genes\")\n",
    "write.csv(as.data.frame(resSig_L10), file=\"CC2344.resSig_L10.genes\")\n",
    "write.csv(as.data.frame(resSig_L11), file=\"CC2344.resSig_L11.genes\")\n",
    "write.csv(as.data.frame(resSig_L12), file=\"CC2344.resSig_L12.genes\")\n",
    "write.csv(as.data.frame(resSig_L13), file=\"CC2344.resSig_L13.genes\")\n",
    "write.csv(as.data.frame(resSig_L14), file=\"CC2344.resSig_L14.genes\")\n",
    "write.csv(as.data.frame(resSig_L15), file=\"CC2344.resSig_L15.genes\")\n",
    "\n",
    "#### EXPORTING LOG2FOLD CHANGE RESULTS ####\n",
    "write.table(as.data.frame(res_L1), 'genes_log2fold/CC2344-L1',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L2), 'genes_log2fold/CC2344-L2',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L3), 'genes_log2fold/CC2344-L3',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L4), 'genes_log2fold/CC2344-L4',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L5), 'genes_log2fold/CC2344-L5',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L6), 'genes_log2fold/CC2344-L6',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L7), 'genes_log2fold/CC2344-L7',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L8), 'genes_log2fold/CC2344-L8',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L9), 'genes_log2fold/CC2344-L9',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L10), 'genes_log2fold/CC2344-L10',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L11), 'genes_log2fold/CC2344-L11',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L12), 'genes_log2fold/CC2344-L12',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L13), 'genes_log2fold/CC2344-L13',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L14), 'genes_log2fold/CC2344-L14',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L15), 'genes_log2fold/CC2344-L15',  sep=\"\\t\", quote = F)\n",
    "\n",
    "# # #PCA PLOT\n",
    "# # vsdata <- vst(dds, blind=FALSE)\n",
    "# # plotPCA(vsdata, intgroup=\"condition\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-personal] *",
   "language": "python",
   "name": "conda-env-.conda-personal-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
